

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Working with Features using the Dseqrecord class &mdash; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=7ab3649f" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=945798ad" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=6efefdf6"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            pydna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Pydna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../module_contents.html">Module contents</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../example_gallery.html">Example gallery</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pydna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Working with Features using the Dseqrecord class</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/markdown_notebooks/Dseq_Features.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="working-with-features-using-the-dseqrecord-class">
<h1>Working with Features using the Dseqrecord class<a class="headerlink" href="#working-with-features-using-the-dseqrecord-class" title="Permalink to this heading"></a></h1>
<blockquote>
<div><p>Before working with features, check how to import sequences from files in the <a class="reference internal" href="#./Importing_Seqs.ipynb"><span class="xref myst">Importing_Seqs notebook</span></a>.</p>
<p>For full library documentation, visit <a class="reference external" href="https://bjornfjohansson.github.io/pydna/">here</a>.</p>
</div></blockquote>
<p>Some sequence file formats (like Genbank) include features, describing key biological properties of sequence regions. In Genbank, features “include genes, gene products, as well as regions of biological significance reported in the sequence.” (See <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/samplerecord/">here</a> for a description of a Genbank file and associated terminologies/annotations) Examples include coding sequences (CDS), introns, promoters, etc.</p>
<p>pydna offers many ways to easily view, add, extract, and write features into a Genbank file via the <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> class. After reading a file into a <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> object, we can check out the list of features in the record using the following code. This example uses the sample record <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/samplerecord/">U49845</a>.</p>
<a target="_blank" href="https://colab.research.google.com/github/BjornFJohansson/pydna/blob/dev_bjorn/docs/notebooks/Dseq_Features.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># Install pydna (only when running on Colab)
import sys
if &#39;google.colab&#39; in sys.modules:
    %%capture
    # Install the current development version of pydna (comment to install pip version)
    !pip install git+https://github.com/BjornFJohansson/pydna@dev_bjorn
    # Install pip version instead (uncomment to install)
    # !pip install pydna

</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="kn">from</span> <span class="nn">pydna.parsers</span> <span class="kn">import</span> <span class="n">parse</span>

<span class="c1">#Import your file into python. </span>
<span class="n">file_path</span> <span class="o">=</span> <span class="s2">&quot;./U49845.gb&quot;</span>
<span class="n">records</span> <span class="o">=</span> <span class="n">parse</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
<span class="n">sample_record</span> <span class="o">=</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># List all features</span>
<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: source
location: [0:5028](+)
qualifiers:
    Key: chromosome, Value: [&#39;IX&#39;]
    Key: db_xref, Value: [&#39;taxon:4932&#39;]
    Key: mol_type, Value: [&#39;genomic DNA&#39;]
    Key: organism, Value: [&#39;Saccharomyces cerevisiae&#39;]

type: mRNA
location: [&lt;0:&gt;206](+)
qualifiers:
    Key: product, Value: [&#39;TCP1-beta&#39;]

type: CDS
location: [&lt;0:206](+)
qualifiers:
    Key: codon_start, Value: [&#39;3&#39;]
    Key: product, Value: [&#39;TCP1-beta&#39;]
    Key: protein_id, Value: [&#39;AAA98665.1&#39;]
    Key: translation, Value: [&#39;SSIYNGISTSGLDLNNGTIADMRQLGIVESYKLKRAVVSSASEAAEVLLRVDNIIRARPRTANRQHM&#39;]

type: gene
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]

type: mRNA
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]
    Key: product, Value: [&#39;Axl2p&#39;]

type: CDS
location: [686:3158](+)
qualifiers:
    Key: codon_start, Value: [&#39;1&#39;]
    Key: gene, Value: [&#39;AXL2&#39;]
    Key: note, Value: [&#39;plasma membrane glycoprotein&#39;]
    Key: product, Value: [&#39;Axl2p&#39;]
    Key: protein_id, Value: [&#39;AAA98666.1&#39;]
    Key: translation, Value: [&#39;MTQLQISLLLTATISLLHLVVATPYEAYPIGKQYPPVARVNESFTFQISNDTYKSSVDKTAQITYNCFDLPSWLSFDSSSRTFSGEPSSDLLSDANTTLYFNVILEGTDSADSTSLNNTYQFVVTNRPSISLSSDFNLLALLKNYGYTNGKNALKLDPNEVFNVTFDRSMFTNEESIVSYYGRSQLYNAPLPNWLFFDSGELKFTGTAPVINSAIAPETSYSFVIIATDIEGFSAVEVEFELVIGAHQLTTSIQNSLIINVTDTGNVSYDLPLNYVYLDDDPISSDKLGSINLLDAPDWVALDNATISGSVPDELLGKNSNPANFSVSIYDTYGDVIYFNFEVVSTTDLFAISSLPNINATRGEWFSYYFLPSQFTDYVNTNVSLEFTNSSQDHDWVKFQSSNLTLAGEVPKNFDKLSLGLKANQGSQSQELYFNIIGMDSKITHSNHSANATSTRSSHHSTSTSSYTSSTYTAKISSTSAAATSSAPAALPAANKTSSHNKKAVAIACGVAIPLGVILVALICFLIFWRRRRENPDDENLPHAISGPDLNNPANKPNQENATPLNNPFDDDASSYDDTSIARRLAALNTLKLDNHSATESDISSVDEKRDSLSGMNTYNDQFQSQSKEELLAKPPVQPPESPFFDPQNRSSSVYMDSEPAVNKSWRYTGNLSPVSDIVRDSYGSQKTVDTEKLFDLEAPEKEKRTSRDVTMSSLDPWNSNISPSPVRKSVTPSPYNVTKHRNRHLQNIQDSQSGKNGITPTTMSTSSSDDFVPVKDGENFCWVHSMEPDRRPSKKRLVDFSNKSNVNVGQVKDIHGRIPEML&#39;]

type: gene
location: [&lt;3299:&gt;4037](-)
qualifiers:
    Key: gene, Value: [&#39;REV7&#39;]

type: mRNA
location: [&lt;3299:&gt;4037](-)
qualifiers:
    Key: gene, Value: [&#39;REV7&#39;]
    Key: product, Value: [&#39;Rev7p&#39;]

type: CDS
location: [3299:4037](-)
qualifiers:
    Key: codon_start, Value: [&#39;1&#39;]
    Key: gene, Value: [&#39;REV7&#39;]
    Key: product, Value: [&#39;Rev7p&#39;]
    Key: protein_id, Value: [&#39;AAA98667.1&#39;]
    Key: translation, Value: [&#39;MNRWVEKWLRVYLKCYINLILFYRNVYPPQSFDYTTYQSFNLPQFVPINRHPALIDYIEELILDVLSKLTHVYRFSICIINKKNDLCIEKYVLDFSELQHVDKDDQIITETEVFDEFRSSLNSLIMHLEKLPKVNDDTITFEAVINAIELELGHKLDRNRRVDSLEEKAEIERDSNWVKCQEDENLPDNNGFQPPKIKLTSLVGSDVGPLIIHQFSEKLISGDDKILNGVYSQYEEGESIFGSLF&#39;]
</pre></div>
</div>
<p>Additional ways to view and search for particular features are shown at the bottom of the page under “Other Methods to Viewing Features”</p>
<section id="adding-features-and-qualifiers">
<h2>Adding Features and Qualifiers<a class="headerlink" href="#adding-features-and-qualifiers" title="Permalink to this heading"></a></h2>
<p>To add new feature to describe a region of interest to a record, for instance a region that you would like to perform a PCR, you need to create a <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code> (sequence feature). The minimal information required is:</p>
<ul class="simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">FeatureLocation</span></code>: position of the feature in the sequence.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">type</span></code> of feature you want to add.</p></li>
</ul>
<p>🚨🚨 <strong>VERY IMPORTANT</strong> 🚨🚨. Note that <code class="docutils literal notranslate"><span class="pre">FeatureLocation</span></code>s are like python ranges (zero-based open intervals), whereas in GenBank files, locations are one-based closed intervals. For instance, the following code adds a new feature from the 2nd to the 5th nucleotide (<code class="docutils literal notranslate"><span class="pre">FeatureLocation(3,</span> <span class="pre">15)</span></code>), of the <code class="docutils literal notranslate"><span class="pre">gene</span></code> type, but in the GenBank file will be represented as <code class="docutils literal notranslate"><span class="pre">4..15</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span><span class="p">,</span> <span class="n">SeqFeature</span>

<span class="c1"># Create a dummy record</span>
<span class="n">dummy_record</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaaATGCGTACGTGAacgt&quot;</span><span class="p">)</span>

<span class="c1"># Define the locations of a CDS</span>
<span class="n">location</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>

<span class="c1"># Create a SeqFeature with the type mRNA</span>
<span class="n">my_feature</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">location</span><span class="o">=</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>

<span class="c1"># Add my_feature to dummy_record with .append</span>
<span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">my_feature</span><span class="p">)</span>

<span class="c1"># Confirm that my_feature has been added</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Print the feature in GenBank format (see how the location is `4..15`)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: gene
location: [3:15]
qualifiers:

LOCUS       name                      19 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     gene            4..15
ORIGIN
        1 aaaatgcgta cgtgaacgt
//
</pre></div>
</div>
<p>To give further information about a feature, we can add a qualifier using the <code class="docutils literal notranslate"><span class="pre">qualifiers</span></code> property of <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code>, which contains a dictionary of qualifiers. For instance, if I would like to note a new feature of type ‘domain’, between 3-9 bases as my region of interest, I can instantiate the <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code> class object as such.</p>
<blockquote>
<div><p>Note that a new feature is always added to the last position of the features list.</p>
</div></blockquote>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">location</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">)</span>

<span class="c1"># Create a SeqFeature with a qualifier</span>
<span class="n">my_feature2</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">location</span><span class="o">=</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Note&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Region of interest&quot;</span><span class="p">]})</span>

<span class="c1"># Add my_feature to my_record with .append</span>
<span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">my_feature2</span><span class="p">)</span>

<span class="c1"># Confirm that my_feature has been added</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;&gt; Feature was added:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">()</span>

<span class="c1"># Print the feature in GenBank format</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;&gt; GenBank format:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>


</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;&gt; Feature was added:
type: domain
location: [3:9]
qualifiers:
    Key: Note, Value: [&#39;Region of interest&#39;]


&gt;&gt; GenBank format:
LOCUS       name                      19 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     gene            4..15
     domain          4..9
                     /Note=&quot;Region of interest&quot;
ORIGIN
        1 aaaatgcgta cgtgaacgt
//
</pre></div>
</div>
<p><strong>🤔 Best practices for qualifiers:</strong></p>
<p>The values in the <code class="docutils literal notranslate"><span class="pre">qualifiers</span></code> dictionary should be lists. The reason for this is that in a GenBank file, a single feature can have multiple values for a single qualifier. Below is a real world of the ase1 CDS example from the <em>S. pombe</em> genome in EMBL format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">FT</span>   <span class="n">CDS</span>             <span class="n">join</span><span class="p">(</span><span class="mf">1878362..1878785</span><span class="p">,</span><span class="mf">1878833..1880604</span><span class="p">)</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">colour</span><span class="o">=</span><span class="mi">2</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">primary_name</span><span class="o">=</span><span class="s2">&quot;ase1&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">product</span><span class="o">=</span><span class="s2">&quot;antiparallel microtubule cross-linking factor</span>
<span class="n">FT</span>                   <span class="n">Ase1</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">systematic_id</span><span class="o">=</span><span class="s2">&quot;SPAPB1A10.09&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution, conserved</span>
<span class="n">FT</span>                   <span class="ow">in</span> <span class="n">eukaryotes</span><span class="p">;</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution, conserved</span>
<span class="n">FT</span>                   <span class="ow">in</span> <span class="n">metazoa</span><span class="p">;</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution, conserved</span>
<span class="n">FT</span>                   <span class="ow">in</span> <span class="n">vertebrates</span><span class="p">;</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution,</span>
<span class="n">FT</span>                   <span class="n">predominantly</span> <span class="n">single</span> <span class="n">copy</span> <span class="p">(</span><span class="n">one</span> <span class="n">to</span> <span class="n">one</span><span class="p">);</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution, conserved</span>
<span class="n">FT</span>                   <span class="ow">in</span> <span class="n">fungi</span><span class="p">;</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
<span class="n">FT</span>                   <span class="o">/</span><span class="n">controlled_curation</span><span class="o">=</span><span class="s2">&quot;term=species distribution, conserved</span>
<span class="n">FT</span>                   <span class="ow">in</span> <span class="n">eukaryotes</span> <span class="n">only</span><span class="p">;</span> <span class="n">date</span><span class="o">=</span><span class="mi">20081110</span><span class="s2">&quot;</span>
</pre></div>
</div>
<p>Note how there are several <code class="docutils literal notranslate"><span class="pre">controlled_curation</span></code> qualifiers, therefore it makes sense to store them as a list.</p>
<p>By default, you can add any type of object in the qualifiers dictionary, and most things will work if you add a string. However, you risk overwriting the existing value for a qualifier, so best practice is:</p>
<ol class="arabic simple">
<li><p>Check if the qualifier already exists using <code class="docutils literal notranslate"><span class="pre">if</span> <span class="pre">&quot;qualifier_name&quot;</span> <span class="pre">in</span> <span class="pre">feature.qualifiers</span></code></p></li>
<li><p>If it exists, append to the existing list of values using <code class="docutils literal notranslate"><span class="pre">feature.qualifiers[&quot;qualifier_name&quot;].append(&quot;new_value&quot;)</span></code></p></li>
<li><p>If it does not exist, add it to the qualifiers dictionary using <code class="docutils literal notranslate"><span class="pre">feature.qualifiers[&quot;qualifier_name&quot;]</span> <span class="pre">=</span> <span class="pre">[&quot;new_value&quot;]</span></code></p></li>
</ol>
<p>Note that <code class="docutils literal notranslate"><span class="pre">Bio.SeqFeatures</span></code> does not automatically assume a sequence strand for the feature. If you would like to refer to a feature on the positive or minus strand, you can add a parameter in <code class="docutils literal notranslate"><span class="pre">FeatureLocation</span></code> specifying <code class="docutils literal notranslate"><span class="pre">strand=+1</span></code> or <code class="docutils literal notranslate"><span class="pre">strand=-1</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#Create a location specifying the minus strand</span>
<span class="n">location</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">19</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

<span class="n">my_feature3</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">location</span><span class="o">=</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:[</span><span class="s2">&quot;example_domain&quot;</span><span class="p">]})</span>

<span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">my_feature3</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>


</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: domain
location: [15:19](-)
qualifiers:
    Key: gene, Value: [&#39;example_domain&#39;]

LOCUS       name                      19 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     gene            4..15
     domain          4..9
                     /Note=&quot;Region of interest&quot;
     domain          complement(16..19)
                     /gene=&quot;example_domain&quot;
ORIGIN
        1 aaaatgcgta cgtgaacgt
//
</pre></div>
</div>
<section id="adding-a-feature-with-parts">
<h3>Adding a Feature with Parts<a class="headerlink" href="#adding-a-feature-with-parts" title="Permalink to this heading"></a></h3>
<p>To add a feature with parts, like a CDS with introns, we need to use a <code class="docutils literal notranslate"><span class="pre">CompoundLocation</span></code> object when creating a <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code>.</p>
<p>The example code below adds a CDS with two parts, between 3-9bp and 12-15bp, to my features list. In a real-world scenario this would represent a CDS with an intron that skips the <code class="docutils literal notranslate"><span class="pre">ACG</span></code> codon: ATGCGT~~ACG~~TGA</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">CompoundLocation</span>

<span class="c1"># Define the locations of the CDS</span>
<span class="n">locations</span> <span class="o">=</span> <span class="p">[</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">15</span><span class="p">)]</span>

<span class="c1"># Create a compound location from these parts</span>
<span class="n">compound_location</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span>

<span class="c1"># Create a SeqFeature with this compound location, including type and qualifiers. </span>
<span class="n">cds_feature</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">location</span><span class="o">=</span><span class="n">compound_location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;example_gene&quot;</span><span class="p">]})</span>

<span class="c1"># Add the feature to the Dseqrecord</span>
<span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds_feature</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dummy_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: CDS
location: join{[3:9], [12:15]}
qualifiers:
    Key: gene, Value: [&#39;example_gene&#39;]

LOCUS       name                      19 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     gene            4..15
     domain          4..9
                     /Note=&quot;Region of interest&quot;
     domain          complement(16..19)
                     /gene=&quot;example_domain&quot;
     CDS             join(4..9,13..15)
                     /gene=&quot;example_gene&quot;
ORIGIN
        1 aaaatgcgta cgtgaacgt
//
</pre></div>
</div>
<p>We can even extract a protein record as follows (see how the protein sequence is <code class="docutils literal notranslate"><span class="pre">MR</span></code>, skipping the intron):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sub_record</span> <span class="o">=</span> <span class="n">dummy_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">dummy_record</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">sub_record</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>ID: id
Name: name
Description: description
Number of features: 0
/molecule_type=DNA
ProteinSeq(&#39;MR&#39;)
</pre></div>
</div>
</section>
<section id="standard-feature-types-and-qualifiers">
<h3>Standard Feature Types and Qualifiers<a class="headerlink" href="#standard-feature-types-and-qualifiers" title="Permalink to this heading"></a></h3>
<p><code class="docutils literal notranslate"><span class="pre">pydna</span></code> and <code class="docutils literal notranslate"><span class="pre">Bio.SeqFeature</span></code> suppports all the conventional feature types through the <code class="docutils literal notranslate"><span class="pre">type</span></code> parameters. A non-exhaustive list include gene, CDS, promoter, exon, intron, 5’ UTR, 3’ UTR, terminator, enhancer, and RBS. You can also define custom features, which could be useful for synthetic biology applications. For instance, you might want to have Bio_brick or spacer features to describe a synthetic standardised plasmid construct.</p>
<p>It is important to note that while <code class="docutils literal notranslate"><span class="pre">pydna</span></code> and <code class="docutils literal notranslate"><span class="pre">Bio.SeqFeature</span></code> does not restrict the feature types you can use, sticking to standard types helps maintain compatibility with other bioinformatics tools and databases. Please refer to the official <a class="reference external" href="https://www.insdc.org/submitting-standards/feature-table/#2">GenBank_Feature_Table</a>, that lists the standard feature types and their associated qualifiers.</p>
<p>Further documentation for <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code>, <code class="docutils literal notranslate"><span class="pre">CompoundLocation</span></code>, and <code class="docutils literal notranslate"><span class="pre">FeatureLocation</span></code> can be found in the <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code> module <a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SeqFeature.html">here</a>.</p>
</section>
<section id="handling-origin-spanning-features">
<h3>Handling Origin Spanning Features<a class="headerlink" href="#handling-origin-spanning-features" title="Permalink to this heading"></a></h3>
<p>An origin spanning feature is a special type of feature that crosses over a circular sequence’s origin. In pydna, such a feature is represented as a feature with parts, joining the part of the sequence immediately before the origin and immediately after the origin. They can be added using <code class="docutils literal notranslate"><span class="pre">CompoundLocation</span></code> as normal.</p>
<p>An origin spanning feature, between base 19 to base 6, in a 25bp long circular sequence, is represented like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">:</span> <span class="n">gene</span> 
<span class="n">location</span><span class="p">:</span> <span class="n">join</span><span class="p">{[</span><span class="mi">19</span><span class="p">:</span><span class="mi">25</span><span class="p">](</span><span class="o">+</span><span class="p">),</span> <span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">](</span><span class="o">+</span><span class="p">)}</span>  
<span class="n">qualifiers</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="n">Value</span><span class="p">:</span> <span class="n">example_gene</span>  
</pre></div>
</div>
<p>This feature will be displayed as a single feature in SnapGene viewer and Benchling, since they support this convention.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">circular_record</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;ACGTGAaaaaaaaaaaaaaATGCGT&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">location</span> <span class="o">=</span> <span class="p">[</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">19</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">)]</span>
<span class="n">ori_feat_location</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">(</span><span class="n">location</span><span class="p">)</span>
<span class="n">ori_feature</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">location</span><span class="o">=</span><span class="n">ori_feat_location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;misc&quot;</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;example origin spanning gene&quot;</span><span class="p">]})</span>
<span class="n">circular_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ori_feature</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;&gt; Feature:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">circular_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Note how the feature sequence is extracted properly across the origin.</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;&gt; Feature sequence:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">circular_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">circular_record</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;&gt; GenBank format:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">circular_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;&gt; Feature:
type: misc
location: join{[19:25], [0:6]}
qualifiers:
    Key: gene, Value: [&#39;example origin spanning gene&#39;]

&gt;&gt; Feature sequence:
ATGCGTACGTGA

&gt;&gt; GenBank format:
LOCUS       name                      25 bp    DNA     circular UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     misc            join(20..25,1..6)
                     /gene=&quot;example origin spanning gene&quot;
ORIGIN
        1 acgtgaaaaa aaaaaaaaaa tgcgt
//
</pre></div>
</div>
</section>
<section id="other-methods-to-viewing-features">
<h3>Other Methods to Viewing Features<a class="headerlink" href="#other-methods-to-viewing-features" title="Permalink to this heading"></a></h3>
<p>pydna also provides the <code class="docutils literal notranslate"><span class="pre">list_features</span></code> method as a simple way to list all the features in a <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">sample_record</span><span class="o">.</span><span class="n">list_features</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>+-----+------------------+-----+-------+-------+------+--------+------+
| Ft# | Label or Note    | Dir | Sta   | End   |  Len | type   | orf? |
+-----+------------------+-----+-------+-------+------+--------+------+
|   0 | nd               | --&gt; | 0     | 5028  | 5028 | source |  no  |
|   1 | nd               | --&gt; | &lt;0    | &gt;206  |  206 | mRNA   |  no  |
|   2 | nd               | --&gt; | &lt;0    | 206   |  206 | CDS    |  no  |
|   3 | nd               | --&gt; | &lt;686  | &gt;3158 | 2472 | gene   | yes  |
|   4 | nd               | --&gt; | &lt;686  | &gt;3158 | 2472 | mRNA   | yes  |
|   5 | N:plasma membran | --&gt; | 686   | 3158  | 2472 | CDS    | yes  |
|   6 | nd               | &lt;-- | &lt;3299 | &gt;4037 |  738 | gene   | yes  |
|   7 | nd               | &lt;-- | &lt;3299 | &gt;4037 |  738 | mRNA   | yes  |
|   8 | nd               | &lt;-- | 3299  | 4037  |  738 | CDS    | yes  |
+-----+------------------+-----+-------+-------+------+--------+------+
</pre></div>
</div>
<p>This method is convenient for checking-out a brief overview of each feature, without reading through an entire sequence record.</p>
<p>Alternatively, we can look for specific features using their qualifiers. For instance:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter based on feature type</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Getting all CDS features:&#39;</span><span class="p">)</span>
<span class="n">cds_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">cds_features</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Getting all CDS features:
type: CDS
location: [&lt;0:206](+)
qualifiers:
    Key: codon_start, Value: [&#39;3&#39;]
    Key: product, Value: [&#39;TCP1-beta&#39;]
    Key: protein_id, Value: [&#39;AAA98665.1&#39;]
    Key: translation, Value: [&#39;SSIYNGISTSGLDLNNGTIADMRQLGIVESYKLKRAVVSSASEAAEVLLRVDNIIRARPRTANRQHM&#39;]

type: CDS
location: [686:3158](+)
qualifiers:
    Key: codon_start, Value: [&#39;1&#39;]
    Key: gene, Value: [&#39;AXL2&#39;]
    Key: note, Value: [&#39;plasma membrane glycoprotein&#39;]
    Key: product, Value: [&#39;Axl2p&#39;]
    Key: protein_id, Value: [&#39;AAA98666.1&#39;]
    Key: translation, Value: [&#39;MTQLQISLLLTATISLLHLVVATPYEAYPIGKQYPPVARVNESFTFQISNDTYKSSVDKTAQITYNCFDLPSWLSFDSSSRTFSGEPSSDLLSDANTTLYFNVILEGTDSADSTSLNNTYQFVVTNRPSISLSSDFNLLALLKNYGYTNGKNALKLDPNEVFNVTFDRSMFTNEESIVSYYGRSQLYNAPLPNWLFFDSGELKFTGTAPVINSAIAPETSYSFVIIATDIEGFSAVEVEFELVIGAHQLTTSIQNSLIINVTDTGNVSYDLPLNYVYLDDDPISSDKLGSINLLDAPDWVALDNATISGSVPDELLGKNSNPANFSVSIYDTYGDVIYFNFEVVSTTDLFAISSLPNINATRGEWFSYYFLPSQFTDYVNTNVSLEFTNSSQDHDWVKFQSSNLTLAGEVPKNFDKLSLGLKANQGSQSQELYFNIIGMDSKITHSNHSANATSTRSSHHSTSTSSYTSSTYTAKISSTSAAATSSAPAALPAANKTSSHNKKAVAIACGVAIPLGVILVALICFLIFWRRRRENPDDENLPHAISGPDLNNPANKPNQENATPLNNPFDDDASSYDDTSIARRLAALNTLKLDNHSATESDISSVDEKRDSLSGMNTYNDQFQSQSKEELLAKPPVQPPESPFFDPQNRSSSVYMDSEPAVNKSWRYTGNLSPVSDIVRDSYGSQKTVDTEKLFDLEAPEKEKRTSRDVTMSSLDPWNSNISPSPVRKSVTPSPYNVTKHRNRHLQNIQDSQSGKNGITPTTMSTSSSDDFVPVKDGENFCWVHSMEPDRRPSKKRLVDFSNKSNVNVGQVKDIHGRIPEML&#39;]

type: CDS
location: [3299:4037](-)
qualifiers:
    Key: codon_start, Value: [&#39;1&#39;]
    Key: gene, Value: [&#39;REV7&#39;]
    Key: product, Value: [&#39;Rev7p&#39;]
    Key: protein_id, Value: [&#39;AAA98667.1&#39;]
    Key: translation, Value: [&#39;MNRWVEKWLRVYLKCYINLILFYRNVYPPQSFDYTTYQSFNLPQFVPINRHPALIDYIEELILDVLSKLTHVYRFSICIINKKNDLCIEKYVLDFSELQHVDKDDQIITETEVFDEFRSSLNSLIMHLEKLPKVNDDTITFEAVINAIELELGHKLDRNRRVDSLEEKAEIERDSNWVKCQEDENLPDNNGFQPPKIKLTSLVGSDVGPLIIHQFSEKLISGDDKILNGVYSQYEEGESIFGSLF&#39;]
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Find a particular feature by its qualifier (e.g. gene name)</span>
<span class="n">rev7_cds_feature</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> 
     <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">and</span>
     <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span> <span class="ow">and</span> <span class="s2">&quot;REV7&quot;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
     <span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">rev7_cds_feature</span><span class="p">)</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: gene
location: [&lt;3299:&gt;4037](-)
qualifiers:
    Key: gene, Value: [&#39;REV7&#39;]
</pre></div>
</div>
<p>If you would like to search for another type of features, simply replace the <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code> with your desired feature type in quotation marks.</p>
</section>
<section id="removing-features">
<h3>Removing Features<a class="headerlink" href="#removing-features" title="Permalink to this heading"></a></h3>
<p>In pydna, we can search for the feature that we would like to remove using the feature’s types or qualififers. For instance, we can modify the features list to exclude all CDS:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#Remove all CDS type features from my feature list</span>
<span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">)]</span>

<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: source
location: [0:5028](+)
qualifiers:
    Key: chromosome, Value: [&#39;IX&#39;]
    Key: db_xref, Value: [&#39;taxon:4932&#39;]
    Key: mol_type, Value: [&#39;genomic DNA&#39;]
    Key: organism, Value: [&#39;Saccharomyces cerevisiae&#39;]

type: mRNA
location: [&lt;0:&gt;206](+)
qualifiers:
    Key: product, Value: [&#39;TCP1-beta&#39;]

type: gene
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]

type: mRNA
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]
    Key: product, Value: [&#39;Axl2p&#39;]

type: gene
location: [&lt;3299:&gt;4037](-)
qualifiers:
    Key: gene, Value: [&#39;REV7&#39;]

type: mRNA
location: [&lt;3299:&gt;4037](-)
qualifiers:
    Key: gene, Value: [&#39;REV7&#39;]
    Key: product, Value: [&#39;Rev7p&#39;]
</pre></div>
</div>
<p>We can also modify the features list to exclude a specific gene:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#Exclude REV7 from my feature list</span>
<span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="s1">&#39;gene&#39;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span> <span class="ow">and</span> <span class="s1">&#39;REV7&#39;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">])]</span>

<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">sample_record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>type: source
location: [0:5028](+)
qualifiers:
    Key: chromosome, Value: [&#39;IX&#39;]
    Key: db_xref, Value: [&#39;taxon:4932&#39;]
    Key: mol_type, Value: [&#39;genomic DNA&#39;]
    Key: organism, Value: [&#39;Saccharomyces cerevisiae&#39;]

type: mRNA
location: [&lt;0:&gt;206](+)
qualifiers:
    Key: product, Value: [&#39;TCP1-beta&#39;]

type: gene
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]

type: mRNA
location: [&lt;686:&gt;3158](+)
qualifiers:
    Key: gene, Value: [&#39;AXL2&#39;]
    Key: product, Value: [&#39;Axl2p&#39;]
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Björn F. Johansson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>