

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Gibson Assembly in pydna &mdash; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=7ab3649f" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=945798ad" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=6efefdf6"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            pydna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Pydna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../module_contents.html">Module contents</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../example_gallery.html">Example gallery</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pydna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Gibson Assembly in pydna</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/markdown_notebooks/Gibson.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gibson-assembly-in-pydna">
<h1>Gibson Assembly in pydna<a class="headerlink" href="#gibson-assembly-in-pydna" title="Permalink to this heading">ÔÉÅ</a></h1>
<blockquote>
<div><p>Visit the full library documentation <a class="reference external" href="https://bjornfjohansson.github.io/pydna/">here</a></p>
</div></blockquote>
<p>Gibson Assembly is a powerful method to assemble multiple DNA fragments into a single, continuous sequence in a seamless, one-step reaction. Developed by Daniel Gibson and colleagues in 2009, this method has been widely applied to work in molecular cloning, biotechnology, and synthetic biology.</p>
<p><code class="docutils literal notranslate"><span class="pre">pydna</span></code> provides the <code class="docutils literal notranslate"><span class="pre">Assembly</span></code> class to simulate the assembly of DNA sequences. Below is an example fpr performing Gibson Assembly with pre-existing DNA fragments, followed by primer design for generating these fragments via the <code class="docutils literal notranslate"><span class="pre">pcr</span></code> method, if needed.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Assembly</span></code> takes the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">frags</span></code>: list of DNA fragments as <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">limit</span></code>: the minimum sequence homology required.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">algorithm</span></code>: the function used to find homology regions between DNA fragments. For Gibson Assembly, we use the <code class="docutils literal notranslate"><span class="pre">terminal_overlap</span></code> function, which finds homology regions only at the terminal regions. By default, the <code class="docutils literal notranslate"><span class="pre">Assembly</span></code> class uses the <code class="docutils literal notranslate"><span class="pre">common_sub_strings</span></code> function to find homology regions, which finds homology anywhere, as it could happen in a homologous recombination event.</p></li>
</ul>
<a target="_blank" href="https://colab.research.google.com/github/BjornFJohansson/pydna/blob/dev_bjorn/docs/notebooks/Gibson.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># Install pydna (only when running on Colab)
import sys
if &#39;google.colab&#39; in sys.modules:
    %%capture
    # Install the current development version of pydna (comment to install pip version)
    !pip install git+https://github.com/BjornFJohansson/pydna@dev_bjorn
    # Install pip version instead (uncomment to install)
    # !pip install pydna

</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="kn">import</span> <span class="n">Assembly</span>
<span class="kn">from</span> <span class="nn">pydna.common_sub_strings</span> <span class="kn">import</span> <span class="n">terminal_overlap</span>

<span class="c1">#Creating example Dseqrecord sequences</span>
<span class="n">fragment1</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="n">fragment2</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="n">fragment3</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>

<span class="c1">#Creating a list of sequences to assemble</span>
<span class="n">fragments</span> <span class="o">=</span> <span class="p">[</span><span class="n">fragment1</span><span class="p">,</span> <span class="n">fragment2</span><span class="p">,</span> <span class="n">fragment3</span><span class="p">]</span>

<span class="c1">#Performing Gibson assembly, with a minimum shared homology of 14bp</span>
<span class="n">assembly</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">(</span><span class="n">fragments</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="n">terminal_overlap</span><span class="p">)</span>

<span class="c1">#Displaying the assembled product</span>
<span class="nb">print</span><span class="p">(</span><span class="n">assembly</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Assembly
fragments..: 33bp 34bp 35bp
limit(bp)..: 14
G.nodes....: 6
algorithm..: terminal_overlap
</pre></div>
</div>
<p>The printed output shows the length of each fragment provided to the assembly, the minimum length required for sequence homology search, the number of nodes (number of overlapping regions), and the algorithm used for sequence homology search. Please refer to the full <code class="docutils literal notranslate"><span class="pre">Assembly</span></code> module documentation for more information on the algorithm applied.</p>
<p>To make a circular sequence from an <code class="docutils literal notranslate"><span class="pre">Assembly</span></code>, pydna provides the <code class="docutils literal notranslate"><span class="pre">assemble_circular</span></code> method. The assembled sequence can be printed as normal, as <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects. Note that the <code class="docutils literal notranslate"><span class="pre">assemble_circular</span></code> method returns a list, where the two elements are reverse complement of each other.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pydna.contig</span> <span class="kn">import</span> <span class="n">Contig</span>

<span class="c1">#Circularizing the assembled sequence</span>
<span class="n">assembly_circ</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>

<span class="c1">#Printing the sequence records</span>
<span class="nb">print</span><span class="p">(</span><span class="n">assembly_circ</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">assembly_circ</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord
circular: True
size: 59
ID: id
Name: name
Description: description
Number of features: 0
/molecule_type=DNA
Dseq(o59)
acga..GGGt
tgct..CCCa

Dseqrecord
circular: True
size: 59
ID: id
Name: name
Description: description
Number of features: 0
/molecule_type=DNA
Dseq(o59)
taga..AAAA
atct..TTTT
</pre></div>
</div>
<p>Please refer to the Example_Gibson page for an example of a completed workflow for modelling Gibson Assembly using pydna.</p>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Bj√∂rn F. Johansson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>