

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>README Example &mdash; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=7ab3649f" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=a051be83" />

  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=6efefdf6"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            pydna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Pydna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/index.html">Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../example_gallery.html">Example gallery</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pydna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">README Example</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/markdown_notebooks/readme_example.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="readme-example">
<h1>README Example<a class="headerlink" href="#readme-example" title="Permalink to this heading">ÔÉÅ</a></h1>
<p>This notebook contains the example shown in the README file.</p>
<a target="_blank" href="https://colab.research.google.com/github/BjornFJohansson/pydna/blob/dev_bjorn/docs/notebooks/Example_Restriction.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># Install pydna (only when running on Colab)
import sys
if &#39;google.colab&#39; in sys.modules:
    %%capture
    # Install the current development version of pydna (comment to install pip version)
    !pip install git+https://github.com/BjornFJohansson/pydna@dev_bjorn
    # Install pip version instead (uncomment to install)
    # !pip install pydna
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="c1"># Let&#39;s create a DNA sequence record, and add a feature to it</span>
<span class="n">dsr</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ATGCAAACAGTAATGATGGATGACATTCAAAGCACTGATTCTATTGCTGAAAAAGATAAT&quot;</span><span class="p">)</span>
<span class="n">dsr</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;my_gene&quot;</span><span class="p">)</span> <span class="c1"># We add a feature to highlight the sequence as a gene</span>
<span class="n">dsr</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(-60)
[48;5;11mATGCAAACAGTAATGATGGATGACATTCAAAGCACTGATTCTATTGCTGAAAAAGATAAT[0m
TACGTTTGTCATTACTACCTACTGTAAGTTTCGTGACTAAGATAACGACTTTTTCTATTA
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This is how it would look as a genbank file</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dsr</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;genbank&quot;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>LOCUS       name                      60 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  description.
ACCESSION   id
VERSION     id
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     misc            1..60
                     /type=&quot;gene&quot;
                     /label=&quot;my_gene&quot;
ORIGIN
        1 atgcaaacag taatgatgga tgacattcaa agcactgatt ctattgctga aaaagataat
//
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s design primers to amplify it</span>
<span class="kn">from</span> <span class="nn">pydna.design</span> <span class="kn">import</span> <span class="n">primer_design</span>
<span class="c1"># limit is the minimum length of the primer, target_tm is the desired melting temperature of the primer</span>
<span class="n">amplicon</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">dsr</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">13</span><span class="p">,</span> <span class="n">target_tm</span><span class="o">=</span><span class="mi">55</span><span class="p">)</span>
<span class="c1"># Let&#39;s print the primers, and a figure that shows where they align with the template sequence</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;forward primer:&quot;</span><span class="p">,</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">forward_primer</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reverse primer:&quot;</span><span class="p">,</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">reverse_primer</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="n">amplicon</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>forward primer: ATGCAAACAGTAATGATGGA
reverse primer: ATTATCTTTTTCAGCAATAGAATCA





5ATGCAAACAGTAATGATGGA...TGATTCTATTGCTGAAAAAGATAAT3
                        |||||||||||||||||||||||||
                       3ACTAAGATAACGACTTTTTCTATTA5
5ATGCAAACAGTAATGATGGA3
 ||||||||||||||||||||
3TACGTTTGTCATTACTACCT...ACTAAGATAACGACTTTTTCTATTA5
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s say we don&#39;t want to just amplify it, but we want to add restriction sites to it!</span>

<span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">pcr</span>
<span class="c1"># We add the restriction sites to the primers</span>
<span class="n">forward_primer</span> <span class="o">=</span> <span class="s2">&quot;ccccGGATCC&quot;</span> <span class="o">+</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="n">reverse_primer</span> <span class="o">=</span> <span class="s2">&quot;ttttGGATCC&quot;</span> <span class="o">+</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">reverse_primer</span>

<span class="c1"># We do the PCR</span>
<span class="n">pcr_product</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">reverse_primer</span><span class="p">,</span> <span class="n">dsr</span><span class="p">)</span>
<span class="c1"># The PCR product is of class `Amplicon`, a subclass of `Dseqrecord`.</span>
<span class="c1"># When doing a figure, it shows where primers anneal.</span>
<span class="n">pcr_product</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>          5ATGCAAACAGTAATGATGGA...TGATTCTATTGCTGAAAAAGATAAT3
                                  |||||||||||||||||||||||||
                                 3ACTAAGATAACGACTTTTTCTATTACCTAGGtttt5
5ccccGGATCCATGCAAACAGTAATGATGGA3
           ||||||||||||||||||||
          3TACGTTTGTCATTACTACCT...ACTAAGATAACGACTTTTTCTATTA5
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># If we want to see the sequence more clearly, we can turn it into a `Dseqrecord`</span>
<span class="n">pcr_product</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="n">pcr_product</span><span class="p">)</span>
<span class="n">pcr_product</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(-80)
ccccGGATCC[48;5;11mATGCAAACAGTAATGATGGATGACATTCAAAGCACTGATTCTATTGCTGAAAAAGATAAT[0mGGATCCaaaa
ggggCCTAGGTACGTTTGTCATTACTACCTACTGTAAGTTTCGTGACTAAGATAACGACTTTTTCTATTACCTAGGtttt
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span> <span class="c1"># cuts GGATCC</span>
<span class="c1"># a, payload, c are the cut fragments</span>
<span class="n">a</span><span class="p">,</span> <span class="n">payload</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="n">pcr_product</span><span class="o">.</span><span class="n">cut</span> <span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="nb">print</span><span class="p">()</span>
<span class="nb">print</span> <span class="p">(</span><span class="n">payload</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="nb">print</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>


</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(-9)
[48;5;11m[0mccccG    
ggggCCTAG

Dseqrecord(-70)
GATCC[48;5;11mATGCAAACAGTAATGATGGATGACATTCAAAGCACTGATTCTATTGCTGAAAAAGATAAT[0mG    
    GTACGTTTGTCATTACTACCTACTGTAAGTTTCGTGACTAAGATAACGACTTTTTCTATTACCTAG

Dseqrecord(-9)
[48;5;11m[0mGATCCaaaa
    Gtttt
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># We create a circular vector to insert the amplicon into</span>
<span class="n">vector</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aatgtttttccctCCCGGGcaaaatAGATCTtgctatgcatcatcgatct&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;vect&quot;</span><span class="p">)</span>
<span class="n">vector</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(o50)
[48;5;11m[0maatgtttttccctCCCGGGcaaaatAGATCTtgctatgcatcatcgatct
ttacaaaaagggaGGGCCCgttttaTCTAGAacgatacgtagtagctaga
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BglII</span> <span class="c1"># cuts AGATCT</span>
<span class="n">linear_vector_bgl</span> <span class="o">=</span> <span class="n">vector</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BglII</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># Linearize the vector at BglII (produces only one fragment)</span>

<span class="c1"># Ligate the fragment of interest to the vector, and call looped() to circularize it</span>
<span class="c1"># synced is used to place the origin coordinate (0) in the same place for rec_vector and vector</span>
<span class="n">rec_vector</span><span class="o">=</span> <span class="p">(</span><span class="n">linear_vector_bgl</span> <span class="o">+</span> <span class="n">payload</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">vector</span><span class="p">)</span>
<span class="n">rec_vector</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(o116)
aatgtttttccctCCCGGGcaaaatAGATCC[48;5;11mATGCAAACAGTAATGATGGATGACATTCAAAGCACTGATTCTATTGCTGAAAAAGATAAT[0mGGATCTtgctatgcatcatcgatct
ttacaaaaagggaGGGCCCgttttaTCTAGGTACGTTTGTCATTACTACCTACTGTAAGTTTCGTGACTAAGATAACGACTTTTTCTATTACCTAGAacgatacgtagtagctaga
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s simulate a Gibson assembly</span>
<span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="kn">import</span> <span class="n">Assembly</span>

<span class="n">fragments</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;aatgtttttccctCACTACGtgctatgcatcat&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment_A&quot;</span><span class="p">),</span>
    <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;tgctatgcatcatCTATGGAcactctaataatg&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment_B&quot;</span><span class="p">),</span>
    <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;cactctaataatgTTACATAaatgtttttccct&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment_C&quot;</span><span class="p">),</span>
<span class="p">]</span>

<span class="c1"># limit is the min. homology length between fragments in the assembly</span>
<span class="n">asm</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">(</span><span class="n">fragments</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># From the assembly object, which can generate all possible products, get a circular</span>
<span class="n">product</span><span class="p">,</span> <span class="o">*</span><span class="n">rest</span> <span class="o">=</span> <span class="n">asm</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>

<span class="c1"># We can print a figure that shows the overlaps between fragments</span>
<span class="n">product</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> -|fragment_A|13
|             \/
|             /\
|             13|fragment_B|13
|                           \/
|                           /\
|                           13|fragment_C|13
|                                         \/
|                                         /\
|                                         13-
|                                            |
 --------------------------------------------
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Or show the final sequence:</span>
<span class="n">Dseqrecord</span><span class="p">(</span><span class="n">product</span><span class="p">)</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Dseqrecord(o60)
[48;5;11m[0maatgtttttccctCACTACGtgctatgcatcatCTATGGAcactctaataatgTTACATA
ttacaaaaagggaGTGATGCacgatacgtagtaGATACCTgtgagattattacAATGTAT
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Bj√∂rn F. Johansson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>