

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pydna.seqrecord &mdash; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=7ab3649f" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=a051be83" />

  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=6efefdf6"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pydna.sequence_picker" href="pydna_sequence_picker.html" />
    <link rel="prev" title="pydna.seq" href="pydna_seq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            pydna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Pydna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="pydna.html">pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna__pretty.html">pydna._pretty</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna__thermodynamic_data.html">pydna._thermodynamic_data</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_all.html">pydna.all</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_amplicon.html">pydna.amplicon</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_amplify.html">pydna.amplify</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_assembly.html">pydna.assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_codon.html">pydna.codon</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_common_sub_strings.html">pydna.common_sub_strings</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_conftest.html">pydna.conftest</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_contig.html">pydna.contig</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_crispr.html">pydna.crispr</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_design.html">pydna.design</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_download.html">pydna.download</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_dseq.html">pydna.dseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_dseqrecord.html">pydna.dseqrecord</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_editor.html">pydna.editor</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_fakeseq.html">pydna.fakeseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_fusionpcr.html">pydna.fusionpcr</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_gateway.html">pydna.gateway</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_gel.html">pydna.gel</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_genbank.html">pydna.genbank</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_genbankfile.html">pydna.genbankfile</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_genbankfixer.html">pydna.genbankfixer</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_genbankrecord.html">pydna.genbankrecord</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_goldengate.html">pydna.goldengate</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_ladders.html">pydna.ladders</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_ligate.html">pydna.ligate</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_myenzymes.html">pydna.myenzymes</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_myprimers.html">pydna.myprimers</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_parsers.html">pydna.parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_primer.html">pydna.primer</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_readers.html">pydna.readers</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_seq.html">pydna.seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">pydna.seqrecord</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pydna.seqrecord.SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord.from_Bio_SeqRecord()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.locus"><code class="docutils literal notranslate"><span class="pre">SeqRecord.locus</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.accession"><code class="docutils literal notranslate"><span class="pre">SeqRecord.accession</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.definition"><code class="docutils literal notranslate"><span class="pre">SeqRecord.definition</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">SeqRecord.reverse_complement()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">SeqRecord.isorf()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.translate"><code class="docutils literal notranslate"><span class="pre">SeqRecord.translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_colors_to_features_for_ape()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.add_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_feature()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.list_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.list_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.extract_feature()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.sorted_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.sorted_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.seguid"><code class="docutils literal notranslate"><span class="pre">SeqRecord.seguid()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.comment"><code class="docutils literal notranslate"><span class="pre">SeqRecord.comment()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.datefunction"><code class="docutils literal notranslate"><span class="pre">SeqRecord.datefunction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.stamp"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stamp()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.lcs"><code class="docutils literal notranslate"><span class="pre">SeqRecord.lcs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.gc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">SeqRecord.cai()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rarecodons()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.startcodon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stopcodon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.express"><code class="docutils literal notranslate"><span class="pre">SeqRecord.express()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.copy"><code class="docutils literal notranslate"><span class="pre">SeqRecord.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.SeqRecord.dump"><code class="docutils literal notranslate"><span class="pre">SeqRecord.dump()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.reverse_complement()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.isorf()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.gc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.cai()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rarecodons()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.startcodon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.stopcodon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.express"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.express()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pydna_sequence_picker.html">pydna.sequence_picker</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_threading_timer_decorator_exit.html">pydna.threading_timer_decorator_exit</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_tm.html">pydna.tm</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_user_cloning.html">pydna.user_cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna_utils.html">pydna.utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../example_gallery.html">Example gallery</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pydna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Modules</a></li>
      <li class="breadcrumb-item active">pydna.seqrecord</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modules/pydna_seqrecord.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="pydna-seqrecord">
<h1>pydna.seqrecord<a class="headerlink" href="#pydna-seqrecord" title="Permalink to this heading"></a></h1>
<span class="target" id="module-pydna.seqrecord"></span><p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></p>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.from_Bio_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_Bio_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SeqRecord</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.from_Bio_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord" title="Permalink to this definition"></a></dt>
<dd><p>Creates a pydnaSeqRecord from a Biopython SeqRecord.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.locus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">locus</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.locus" title="Permalink to this definition"></a></dt>
<dd><p>Alias for name property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.accession">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">accession</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.accession" title="Permalink to this definition"></a></dt>
<dd><p>Alias for id property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.definition">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">definition</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.definition" title="Permalink to this definition"></a></dt>
<dd><p>Alias for description property.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.reverse_complement" title="Permalink to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rc" title="Permalink to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.isorf" title="Permalink to this definition"></a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id2" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.translate" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape">
<span class="sig-name descname"><span class="pre">add_colors_to_features_for_ape</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.add_colors_to_features_for_ape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape" title="Permalink to this definition"></a></dt>
<dd><p>Assign colors to features.</p>
<p>compatible with
the <a class="reference external" href="http://jorgensen.biology.utah.edu/wayned/ape/">ApE editor</a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_feature" title="Permalink to this definition"></a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2),</span>
<span class="go">                           ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;,</span>
<span class="go">            qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.list_features">
<span class="sig-name descname"><span class="pre">list_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.list_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.list_features" title="Permalink to this definition"></a></dt>
<dd><p>Print ASCII table with all features.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">list_features</span><span class="p">())</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">| Ft# | Label or Note | Dir | Sta | End | Len | type | orf? |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">|   0 | L:ft2         | --&gt; | 2   | 4   |   2 | misc |  no  |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.extract_feature" title="Permalink to this definition"></a></dt>
<dd><p>Extract feature and return a new SeqRecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – </p></li>
<li><p><strong>extract</strong> (<em>Indicates the feature to</em>) – </p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">SeqRecord(seq=Seq(&#39;gt&#39;), id=&#39;ft2&#39;, name=&#39;part_name&#39;,</span>
<span class="go">          description=&#39;description&#39;, dbxrefs=[])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.sorted_features">
<span class="sig-name descname"><span class="pre">sorted_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.sorted_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.sorted_features" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of the features sorted by start position.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">sorted_features</span><span class="p">())</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.seguid" title="Permalink to this definition"></a></dt>
<dd><p>Return the url safe SEGUID <a class="footnote-reference brackets" href="#id4" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a> for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base 64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be a part of and URL or a filename.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;gattaca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span> <span class="c1"># original seguid is +bKGnebMkia5kNg/gF7IORXMnIU</span>
<span class="go">&#39;lsseguid=tp2jzeCM2e3W4yxtrrx09CMKa_8&#39;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.comment">
<span class="sig-name descname"><span class="pre">comment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">newcomment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.comment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.comment" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.datefunction">
<span class="sig-name descname"><span class="pre">datefunction</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.datefunction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.datefunction" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stamp">
<span class="sig-name descname"><span class="pre">stamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">now</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">datefunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pydna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'</span> <span class="pre">'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.stamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stamp" title="Permalink to this definition"></a></dt>
<dd><p>Add seguid checksum to COMMENTS sections</p>
<p>The checksum is stored in object.annotations[“comment”].
This shows in the COMMENTS section of a formatted genbank file.</p>
<p>For blunt linear sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">SEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>For circular sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">cSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>Fore linear sequences which are not blunt:</p>
<p><code class="docutils literal notranslate"><span class="pre">lSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="go">&#39;lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttjU&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">][:</span><span class="mi">41</span><span class="p">]</span>
<span class="go">&#39;pydna lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttj&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.lcs">
<span class="sig-name descname"><span class="pre">lcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.lcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.lcs" title="Permalink to this definition"></a></dt>
<dd><p>Return the longest common substring between the sequence.</p>
<p>and another sequence (other). The other sequence can be a string,
Seq, SeqRecord, Dseq or DseqRecord.
The method returns a SeqFeature with type “read” as this method
is mostly used to map sequence reads to the sequence. This can be
changed by passing a type as keyword with some other string value.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(0),</span>
<span class="go">                          ExactPosition(6), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(1),</span>
<span class="go">                          ExactPosition(5), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;CCCCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(None)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.gc" title="Permalink to this definition"></a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.cai" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rarecodons" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.startcodon" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stopcodon" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.express" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.copy" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protocol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.dump" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">ProteinSeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement" title="Permalink to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rc" title="Permalink to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.isorf" title="Permalink to this definition"></a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id6" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">3</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.gc" title="Permalink to this definition"></a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.cai" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.startcodon" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.express" title="Permalink to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pydna_seq.html" class="btn btn-neutral float-left" title="pydna.seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pydna_sequence_picker.html" class="btn btn-neutral float-right" title="pydna.sequence_picker" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Björn F. Johansson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>