<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Welcome to pydna’s documentation! &mdash; pydna 0.0.0.post1+a629c82 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=c56941e8"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            pydna
          </a>
              <div class="version">
                0.0.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to pydna’s documentation!</a><ul>
<li><a class="reference internal" href="#module-pydna">Module contents</a><ul>
<li><a class="reference internal" href="#pydna">pydna</a><ul>
<li><a class="reference internal" href="#pydna-package-layout">pydna package layout</a></li>
<li><a class="reference internal" href="#how-to-use-the-documentation">How to use the documentation</a></li>
<li><a class="reference internal" href="#pydna-source-code">pydna source code</a></li>
<li><a class="reference internal" href="#how-to-get-more-help">How to get more help</a></li>
<li><a class="reference internal" href="#examples-of-pydna-in-use">Examples of pydna in use</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.open_current_folder"><code class="docutils literal notranslate"><span class="pre">open_current_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.open_cache_folder"><code class="docutils literal notranslate"><span class="pre">open_cache_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.open_config_folder"><code class="docutils literal notranslate"><span class="pre">open_config_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.open_log_folder"><code class="docutils literal notranslate"><span class="pre">open_log_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.get_env"><code class="docutils literal notranslate"><span class="pre">get_env()</span></code></a></li>
<li><a class="reference internal" href="#pydna.logo"><code class="docutils literal notranslate"><span class="pre">logo()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.dseq">pydna.dseq module</a><ul>
<li><a class="reference internal" href="#pydna.dseq.Dseq"><code class="docutils literal notranslate"><span class="pre">Dseq</span></code></a><ul>
<li><a class="reference internal" href="#pydna.dseq.Dseq.trunc"><code class="docutils literal notranslate"><span class="pre">Dseq.trunc</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.quick"><code class="docutils literal notranslate"><span class="pre">Dseq.quick()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.from_string"><code class="docutils literal notranslate"><span class="pre">Dseq.from_string()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.from_representation"><code class="docutils literal notranslate"><span class="pre">Dseq.from_representation()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.from_full_sequence_and_overhangs"><code class="docutils literal notranslate"><span class="pre">Dseq.from_full_sequence_and_overhangs()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.mw"><code class="docutils literal notranslate"><span class="pre">Dseq.mw()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.upper"><code class="docutils literal notranslate"><span class="pre">Dseq.upper()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.lower"><code class="docutils literal notranslate"><span class="pre">Dseq.lower()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.find"><code class="docutils literal notranslate"><span class="pre">Dseq.find()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseq.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.rc"><code class="docutils literal notranslate"><span class="pre">Dseq.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.shifted"><code class="docutils literal notranslate"><span class="pre">Dseq.shifted()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.looped"><code class="docutils literal notranslate"><span class="pre">Dseq.looped()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseq.tolinear()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.five_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.five_prime_end()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.three_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.three_prime_end()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.watson_ovhg"><code class="docutils literal notranslate"><span class="pre">Dseq.watson_ovhg()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.fill_in"><code class="docutils literal notranslate"><span class="pre">Dseq.fill_in()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.transcribe"><code class="docutils literal notranslate"><span class="pre">Dseq.transcribe()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.translate"><code class="docutils literal notranslate"><span class="pre">Dseq.translate()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.mung"><code class="docutils literal notranslate"><span class="pre">Dseq.mung()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.T4"><code class="docutils literal notranslate"><span class="pre">Dseq.T4()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.t4"><code class="docutils literal notranslate"><span class="pre">Dseq.t4()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.exo1_front"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_front()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.exo1_end"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_end()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.no_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.unique_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.once_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.twice_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.n_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.seguid"><code class="docutils literal notranslate"><span class="pre">Dseq.seguid()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.isblunt"><code class="docutils literal notranslate"><span class="pre">Dseq.isblunt()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.cas9"><code class="docutils literal notranslate"><span class="pre">Dseq.cas9()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseq.terminal_transferase()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.cut"><code class="docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.cutsite_is_valid"><code class="docutils literal notranslate"><span class="pre">Dseq.cutsite_is_valid()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.get_cutsites"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsites()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.left_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.left_end_position()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.right_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.right_end_position()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.get_cut_parameters"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cut_parameters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseq.apply_cut()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseq.Dseq.get_cutsite_pairs"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsite_pairs()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.dseqrecord">pydna.dseqrecord module</a><ul>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a><ul>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.from_string"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_string()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.circular"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.circular</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.m"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.m()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.extract_feature()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.add_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.add_feature()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.seguid"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.seguid()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.looped"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.looped()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.tolinear()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.terminal_transferase()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.format"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.format()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.write"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.write()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.find"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aminoacids()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.find_aa"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aa()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.map_trace_files"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.map_trace_files()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.linearize"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.linearize()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.no_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.unique_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.once_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.twice_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.n_cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cutters()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.number_of_cuts()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.cas9"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cas9()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.rc"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.synced"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.synced()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.upper"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.upper()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.lower"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.lower()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.orfs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.orfs_to_features"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs_to_features()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_gb_to_clipboard()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_fasta_to_clipboard()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.figure"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.figure()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.shifted"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.shifted()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cut()</span></code></a></li>
<li><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.apply_cut()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.amplicon">pydna.amplicon module</a><ul>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon"><code class="docutils literal notranslate"><span class="pre">Amplicon</span></code></a><ul>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Amplicon.from_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Amplicon.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.rc"><code class="docutils literal notranslate"><span class="pre">Amplicon.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.figure"><code class="docutils literal notranslate"><span class="pre">Amplicon.figure()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.set_forward_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_forward_primer_footprint()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.set_reverse_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_reverse_primer_footprint()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.program"><code class="docutils literal notranslate"><span class="pre">Amplicon.program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.dbd_program"><code class="docutils literal notranslate"><span class="pre">Amplicon.dbd_program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplicon.Amplicon.primers"><code class="docutils literal notranslate"><span class="pre">Amplicon.primers()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.amplify">pydna.amplify module</a><ul>
<li><a class="reference internal" href="#pydna.amplify.Anneal"><code class="docutils literal notranslate"><span class="pre">Anneal</span></code></a><ul>
<li><a class="reference internal" href="#pydna.amplify.Anneal.forward_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.forward_primers</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplify.Anneal.reverse_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.reverse_primers</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplify.Anneal.template"><code class="docutils literal notranslate"><span class="pre">Anneal.template</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplify.Anneal.limit"><code class="docutils literal notranslate"><span class="pre">Anneal.limit</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplify.Anneal.products"><code class="docutils literal notranslate"><span class="pre">Anneal.products</span></code></a></li>
<li><a class="reference internal" href="#pydna.amplify.Anneal.report"><code class="docutils literal notranslate"><span class="pre">Anneal.report()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.amplify.pcr"><code class="docutils literal notranslate"><span class="pre">pcr()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.assembly">pydna.assembly module</a><ul>
<li><a class="reference internal" href="#pydna.assembly.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul>
<li><a class="reference internal" href="#pydna.assembly.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
<li><a class="reference internal" href="#pydna.assembly.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.common_sub_strings">pydna.common_sub_strings module</a><ul>
<li><a class="reference internal" href="#pydna.common_sub_strings.common_sub_strings"><code class="docutils literal notranslate"><span class="pre">common_sub_strings()</span></code></a></li>
<li><a class="reference internal" href="#pydna.common_sub_strings.terminal_overlap"><code class="docutils literal notranslate"><span class="pre">terminal_overlap()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.contig">pydna.contig module</a><ul>
<li><a class="reference internal" href="#pydna.contig.Contig"><code class="docutils literal notranslate"><span class="pre">Contig</span></code></a><ul>
<li><a class="reference internal" href="#pydna.contig.Contig.from_string"><code class="docutils literal notranslate"><span class="pre">Contig.from_string()</span></code></a></li>
<li><a class="reference internal" href="#pydna.contig.Contig.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Contig.from_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.contig.Contig.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Contig.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.contig.Contig.rc"><code class="docutils literal notranslate"><span class="pre">Contig.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.contig.Contig.detailed_figure"><code class="docutils literal notranslate"><span class="pre">Contig.detailed_figure()</span></code></a></li>
<li><a class="reference internal" href="#pydna.contig.Contig.figure"><code class="docutils literal notranslate"><span class="pre">Contig.figure()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.design">pydna.design module</a><ul>
<li><a class="reference internal" href="#pydna.design.primer_design"><code class="docutils literal notranslate"><span class="pre">primer_design()</span></code></a></li>
<li><a class="reference internal" href="#pydna.design.assembly_fragments"><code class="docutils literal notranslate"><span class="pre">assembly_fragments()</span></code></a></li>
<li><a class="reference internal" href="#pydna.design.circular_assembly_fragments"><code class="docutils literal notranslate"><span class="pre">circular_assembly_fragments()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.download">pydna.download module</a></li>
<li><a class="reference internal" href="#module-pydna.editor">pydna.editor module</a><ul>
<li><a class="reference internal" href="#pydna.editor.Editor"><code class="docutils literal notranslate"><span class="pre">Editor</span></code></a><ul>
<li><a class="reference internal" href="#pydna.editor.Editor.open"><code class="docutils literal notranslate"><span class="pre">Editor.open()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.editor.ape"><code class="docutils literal notranslate"><span class="pre">ape()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.gel">pydna.gel module</a><ul>
<li><a class="reference internal" href="#pydna.gel.interpolator"><code class="docutils literal notranslate"><span class="pre">interpolator()</span></code></a></li>
<li><a class="reference internal" href="#pydna.gel.gel"><code class="docutils literal notranslate"><span class="pre">gel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.genbank">pydna.genbank module</a><ul>
<li><a class="reference internal" href="#pydna.genbank.Genbank"><code class="docutils literal notranslate"><span class="pre">Genbank</span></code></a><ul>
<li><a class="reference internal" href="#pydna.genbank.Genbank.nucleotide"><code class="docutils literal notranslate"><span class="pre">Genbank.nucleotide()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.genbank.genbank"><code class="docutils literal notranslate"><span class="pre">genbank()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.genbankfile">pydna.genbankfile module</a><ul>
<li><a class="reference internal" href="#pydna.genbankfile.GenbankFile"><code class="docutils literal notranslate"><span class="pre">GenbankFile</span></code></a><ul>
<li><a class="reference internal" href="#pydna.genbankfile.GenbankFile.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankFile.from_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfile.GenbankFile.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankFile.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfile.GenbankFile.rc"><code class="docutils literal notranslate"><span class="pre">GenbankFile.rc()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.genbankfixer">pydna.genbankfixer module</a><ul>
<li><a class="reference internal" href="#pydna.genbankfixer.parseGBLoc"><code class="docutils literal notranslate"><span class="pre">parseGBLoc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.strip_multiline"><code class="docutils literal notranslate"><span class="pre">strip_multiline()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.toInt"><code class="docutils literal notranslate"><span class="pre">toInt()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.strip_indent"><code class="docutils literal notranslate"><span class="pre">strip_indent()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.concat_dict"><code class="docutils literal notranslate"><span class="pre">concat_dict()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.toJSON"><code class="docutils literal notranslate"><span class="pre">toJSON()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.wrapstring"><code class="docutils literal notranslate"><span class="pre">wrapstring()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.locstr"><code class="docutils literal notranslate"><span class="pre">locstr()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.originstr"><code class="docutils literal notranslate"><span class="pre">originstr()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.toGB"><code class="docutils literal notranslate"><span class="pre">toGB()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankfixer.gbtext_clean"><code class="docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.genbankrecord">pydna.genbankrecord module</a><ul>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord</span></code></a><ul>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.from_string"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_string()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.rc"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.pydna_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.pydna_code()</span></code></a></li>
<li><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord.biopython_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.biopython_code()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.myprimers">pydna.myprimers module</a><ul>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList"><code class="docutils literal notranslate"><span class="pre">PrimerList</span></code></a><ul>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList.accessed"><code class="docutils literal notranslate"><span class="pre">PrimerList.accessed</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList.assign_numbers"><code class="docutils literal notranslate"><span class="pre">PrimerList.assign_numbers()</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList.pydna_code_from_list"><code class="docutils literal notranslate"><span class="pre">PrimerList.pydna_code_from_list()</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList.open_folder"><code class="docutils literal notranslate"><span class="pre">PrimerList.open_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.PrimerList.code"><code class="docutils literal notranslate"><span class="pre">PrimerList.code()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.myprimers.check_primer_numbers"><code class="docutils literal notranslate"><span class="pre">check_primer_numbers()</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.undefined_sequence"><code class="docutils literal notranslate"><span class="pre">undefined_sequence()</span></code></a></li>
<li><a class="reference internal" href="#pydna.myprimers.find_duplicate_primers"><code class="docutils literal notranslate"><span class="pre">find_duplicate_primers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.parsers">pydna.parsers module</a><ul>
<li><a class="reference internal" href="#pydna.parsers.extract_from_text"><code class="docutils literal notranslate"><span class="pre">extract_from_text()</span></code></a></li>
<li><a class="reference internal" href="#pydna.parsers.embl_gb_fasta"><code class="docutils literal notranslate"><span class="pre">embl_gb_fasta()</span></code></a></li>
<li><a class="reference internal" href="#pydna.parsers.parse"><code class="docutils literal notranslate"><span class="pre">parse()</span></code></a></li>
<li><a class="reference internal" href="#pydna.parsers.parse_primers"><code class="docutils literal notranslate"><span class="pre">parse_primers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.primer">pydna.primer module</a><ul>
<li><a class="reference internal" href="#pydna.primer.Primer"><code class="docutils literal notranslate"><span class="pre">Primer</span></code></a><ul>
<li><a class="reference internal" href="#pydna.primer.Primer.footprint"><code class="docutils literal notranslate"><span class="pre">Primer.footprint</span></code></a></li>
<li><a class="reference internal" href="#pydna.primer.Primer.tail"><code class="docutils literal notranslate"><span class="pre">Primer.tail</span></code></a></li>
<li><a class="reference internal" href="#pydna.primer.Primer.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Primer.reverse_complement()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.readers">pydna.readers module</a><ul>
<li><a class="reference internal" href="#pydna.readers.read"><code class="docutils literal notranslate"><span class="pre">read()</span></code></a></li>
<li><a class="reference internal" href="#pydna.readers.read_primer"><code class="docutils literal notranslate"><span class="pre">read_primer()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.seqrecord">pydna.seqrecord module</a><ul>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code></a><ul>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord.from_Bio_SeqRecord()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.locus"><code class="docutils literal notranslate"><span class="pre">SeqRecord.locus</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.accession"><code class="docutils literal notranslate"><span class="pre">SeqRecord.accession</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.definition"><code class="docutils literal notranslate"><span class="pre">SeqRecord.definition</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">SeqRecord.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">SeqRecord.isorf()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.translate"><code class="docutils literal notranslate"><span class="pre">SeqRecord.translate()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_colors_to_features_for_ape()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.add_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_feature()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.list_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.list_features()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.extract_feature()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.sorted_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.sorted_features()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.seguid"><code class="docutils literal notranslate"><span class="pre">SeqRecord.seguid()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.comment"><code class="docutils literal notranslate"><span class="pre">SeqRecord.comment()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.datefunction"><code class="docutils literal notranslate"><span class="pre">SeqRecord.datefunction()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.stamp"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stamp()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.lcs"><code class="docutils literal notranslate"><span class="pre">SeqRecord.lcs()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.gc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">SeqRecord.cai()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rarecodons()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.startcodon()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stopcodon()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.express"><code class="docutils literal notranslate"><span class="pre">SeqRecord.express()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.copy"><code class="docutils literal notranslate"><span class="pre">SeqRecord.copy()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.SeqRecord.dump"><code class="docutils literal notranslate"><span class="pre">SeqRecord.dump()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord</span></code></a><ul>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.reverse_complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.isorf()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.gc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.cai()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rarecodons()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.startcodon()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.stopcodon()</span></code></a></li>
<li><a class="reference internal" href="#pydna.seqrecord.ProteinSeqRecord.express"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.express()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.tm">pydna.tm module</a><ul>
<li><a class="reference internal" href="#pydna.tm.tm_default"><code class="docutils literal notranslate"><span class="pre">tm_default()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.tm_dbd"><code class="docutils literal notranslate"><span class="pre">tm_dbd()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.tm_product"><code class="docutils literal notranslate"><span class="pre">tm_product()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.ta_default"><code class="docutils literal notranslate"><span class="pre">ta_default()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.ta_dbd"><code class="docutils literal notranslate"><span class="pre">ta_dbd()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.program"><code class="docutils literal notranslate"><span class="pre">program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.taq_program"><code class="docutils literal notranslate"><span class="pre">taq_program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.dbd_program"><code class="docutils literal notranslate"><span class="pre">dbd_program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.pfu_sso7d_program"><code class="docutils literal notranslate"><span class="pre">pfu_sso7d_program()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.Q5"><code class="docutils literal notranslate"><span class="pre">Q5()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.tmbresluc"><code class="docutils literal notranslate"><span class="pre">tmbresluc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.tm.tm_neb"><code class="docutils literal notranslate"><span class="pre">tm_neb()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.utils">pydna.utils module</a><ul>
<li><a class="reference internal" href="#pydna.utils.three_frame_orfs"><code class="docutils literal notranslate"><span class="pre">three_frame_orfs()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.shift_location"><code class="docutils literal notranslate"><span class="pre">shift_location()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.shift_feature"><code class="docutils literal notranslate"><span class="pre">shift_feature()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.smallest_rotation"><code class="docutils literal notranslate"><span class="pre">smallest_rotation()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.cai"><code class="docutils literal notranslate"><span class="pre">cai()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.rarecodons"><code class="docutils literal notranslate"><span class="pre">rarecodons()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.express"><code class="docutils literal notranslate"><span class="pre">express()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.open_folder"><code class="docutils literal notranslate"><span class="pre">open_folder()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.rc"><code class="docutils literal notranslate"><span class="pre">rc()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.complement"><code class="docutils literal notranslate"><span class="pre">complement()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.memorize"><code class="docutils literal notranslate"><span class="pre">memorize()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.identifier_from_string"><code class="docutils literal notranslate"><span class="pre">identifier_from_string()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.flatten"><code class="docutils literal notranslate"><span class="pre">flatten()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.seq31"><code class="docutils literal notranslate"><span class="pre">seq31()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.randomRNA"><code class="docutils literal notranslate"><span class="pre">randomRNA()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.randomDNA"><code class="docutils literal notranslate"><span class="pre">randomDNA()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.randomORF"><code class="docutils literal notranslate"><span class="pre">randomORF()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.randomprot"><code class="docutils literal notranslate"><span class="pre">randomprot()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.eq"><code class="docutils literal notranslate"><span class="pre">eq()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.cuts_overlap"><code class="docutils literal notranslate"><span class="pre">cuts_overlap()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.location_boundaries"><code class="docutils literal notranslate"><span class="pre">location_boundaries()</span></code></a></li>
<li><a class="reference internal" href="#pydna.utils.locations_overlap"><code class="docutils literal notranslate"><span class="pre">locations_overlap()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">pydna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Welcome to pydna’s documentation!</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="welcome-to-pydna-s-documentation">
<h1>Welcome to pydna’s documentation!<a class="headerlink" href="#welcome-to-pydna-s-documentation" title="Link to this heading"></a></h1>
<p>Stuff &amp; other stuff</p>
<section id="module-pydna">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pydna" title="Link to this heading"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">copyright<span class="colon">:</span></dt>
<dd class="field-odd"><p>Copyright 2013-2023 by Björn Johansson. All rights reserved.</p>
</dd>
<dt class="field-even">license<span class="colon">:</span></dt>
<dd class="field-even"><p>This code is part of the pydna package, governed by the
license in LICENSE.txt that should be included as part
of this package.</p>
</dd>
</dl>
<section id="pydna">
<h3>pydna<a class="headerlink" href="#pydna" title="Link to this heading"></a></h3>
<p>Pydna is a python package providing code for simulation of the creation of
recombinant DNA molecules using
<a class="reference external" href="https://en.wikipedia.org/wiki/Molecular_biology">molecular biology</a>
techniques. Development of pydna happens in this Github <a class="reference external" href="https://github.com/BjornFJohansson/pydna">repository</a>.</p>
<dl class="simple">
<dt>Provided:</dt><dd><ol class="arabic simple">
<li><p>PCR simulation</p></li>
<li><p>Assembly simulation based on shared identical sequences</p></li>
<li><p>Primer design for amplification of a given sequence</p></li>
<li><p>Automatic design of primer tails for Gibson assembly
or homologous recombination.</p></li>
<li><p>Restriction digestion and cut&amp;paste cloning</p></li>
<li><p>Agarose gel simulation</p></li>
<li><p>Download sequences from Genbank</p></li>
<li><p>Parsing various sequence formats including the capacity to
handle broken Genbank format</p></li>
</ol>
</dd>
</dl>
<section id="pydna-package-layout">
<h4>pydna package layout<a class="headerlink" href="#pydna-package-layout" title="Link to this heading"></a></h4>
<p>The most important modules and how to import functions or classes from
them are listed below. Class names starts with a capital letter,
functions with a lowercase letter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>from pydna.module import function
from pydna.module import Class

Example: from pydna.gel import Gel

pydna
   ├── amplify
   │         ├── Anneal
   │         └── pcr
   ├── assembly
   │          └── Assembly
   ├── design
   │        ├── assembly_fragments
   │        └── primer_design
   ├── download
   │          └── download_text
   ├── dseqrecord
   │            └── Dseqrecord
   ├── gel
   │     └── Gel
   ├── genbank
   │         ├── genbank
   │         └── Genbank
   ├── parsers
   │         ├── parse
   │         └── parse_primers
   └── readers
             ├── read
             └── read_primers
</pre></div>
</div>
</section>
<section id="how-to-use-the-documentation">
<h4>How to use the documentation<a class="headerlink" href="#how-to-use-the-documentation" title="Link to this heading"></a></h4>
<p>Documentation is available as docstrings provided in the source code for
each module.
These docstrings can be inspected by reading the source code directly.
See further below on how to obtain the code for pydna.</p>
<p>In the python shell, use the built-in <code class="docutils literal notranslate"><span class="pre">help</span></code> function to view a
function’s docstring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">readers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">readers</span><span class="o">.</span><span class="n">read</span><span class="p">)</span>
<span class="gp">... </span>
</pre></div>
</div>
<p>The doctrings are also used to provide an automaticly generated reference
manual available online at
<a class="reference external" href="https://pydna.readthedocs.io">read the docs</a>.</p>
<p>Docstrings can be explored using <a class="reference external" href="http://ipython.org/">IPython</a>, an
advanced Python shell with
TAB-completion and introspection capabilities. To see which functions
are available in <cite>pydna</cite>,
type <cite>pydna.&lt;TAB&gt;</cite> (where <cite>&lt;TAB&gt;</cite> refers to the TAB key).
Use <cite>pydna.open_config_folder?&lt;ENTER&gt;`to view the docstring or
`pydna.open_config_folder??&lt;ENTER&gt;</cite> to view the source code.</p>
<p>In the <a class="reference external" href="https://github.com/spyder-ide/spyder">Spyder IDE</a> it is possible
to place the cursor immediately before the name of a module,class or
function and press ctrl+i to bring up docstrings in a separate window in Spyder</p>
<p>Code snippets are indicated by three greater-than signs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="mi">41</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">42</span>
</pre></div>
</div>
</section>
<section id="pydna-source-code">
<h4>pydna source code<a class="headerlink" href="#pydna-source-code" title="Link to this heading"></a></h4>
<p>The pydna source code is
<a class="reference external" href="https://github.com/BjornFJohansson/pydna">available on Github</a>.</p>
</section>
<section id="how-to-get-more-help">
<h4>How to get more help<a class="headerlink" href="#how-to-get-more-help" title="Link to this heading"></a></h4>
<p>Please join the
<a class="reference external" href="https://groups.google.com/forum/#!forum/pydna">Google group</a>
for pydna, this is the preferred location for help. If you find bugs
in pydna itself, open an issue at the
<a class="reference external" href="https://github.com/BjornFJohansson/pydna/issues">Github repository</a>.</p>
</section>
<section id="examples-of-pydna-in-use">
<h4>Examples of pydna in use<a class="headerlink" href="#examples-of-pydna-in-use" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>See this repository for a collection of</dt><dd><p><a class="reference external" href="https://github.com/MetabolicEngineeringGroupCBMA/pydna-examples?tab=readme-ov-file#pydna-examples">examples</a>.</p>
</dd>
</dl>
</section>
</section>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.open_current_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">open_current_folder</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_current_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.open_current_folder" title="Link to this definition"></a></dt>
<dd><p>Open the current working directory.</p>
<p>Opens in the default file manager. The location for this folder is
given by the <a class="reference external" href="https://docs.python.org/3.8/library/os.html#os.getcwd" title="(in Python v3.8)"><code class="xref py py-func docutils literal notranslate"><span class="pre">os.getcwd()</span></code></a> function</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.open_cache_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">open_cache_folder</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_cache_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.open_cache_folder" title="Link to this definition"></a></dt>
<dd><p>Open the pydna cache folder.</p>
<p>Opens in the default file manager. The location for this folder is stored
in the <em>pydna_data_dir</em> environmental variable.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.open_config_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">open_config_folder</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_config_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.open_config_folder" title="Link to this definition"></a></dt>
<dd><p>Open the pydna configuration folder.</p>
<p>Opens in the default file manager. The location for this folder is stored
in the <em>pydna_config_dir</em> environmental variable.</p>
<p>The <cite>pydna.ini</cite> file can be edited to make pydna quicker to use.
See the documentation of the :class:configparser.ConfigParser´ class.</p>
<p>Below is the content of a typical <cite>pydna.ini</cite> file on a Linux
system.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">main</span><span class="p">]</span>
<span class="n">loglevel</span><span class="o">=</span><span class="mi">30</span>
<span class="n">email</span><span class="o">=</span><span class="n">myemail</span><span class="nd">@example</span><span class="o">.</span><span class="n">org</span>
<span class="n">data_dir</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">local</span><span class="o">/</span><span class="n">share</span><span class="o">/</span><span class="n">pydna</span>
<span class="n">log_dir</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">cache</span><span class="o">/</span><span class="n">pydna</span><span class="o">/</span><span class="n">log</span>
<span class="n">ape</span><span class="o">=</span><span class="n">tclsh</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">ApE</span><span class="o">/</span><span class="n">AppMain</span><span class="o">.</span><span class="n">tcl</span>
<span class="n">cached_funcs</span><span class="o">=</span><span class="n">Genbank_nucleotide</span>
<span class="n">primers</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/</span><span class="n">Dropbox</span><span class="o">/</span><span class="n">wikidata</span><span class="o">/</span><span class="n">PRIMERS</span><span class="o">.</span><span class="n">txt</span>
<span class="n">enzymes</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/</span><span class="n">Dropbox</span><span class="o">/</span><span class="n">wikidata</span><span class="o">/</span><span class="n">RestrictionEnzymes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The email address is set to <a class="reference external" href="mailto:someone&#37;&#52;&#48;example&#46;com">someone<span>&#64;</span>example<span>&#46;</span>com</a> by default. If you change
this to you own address, the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function can be
used to download sequences from Genbank directly without having to
explicitly add the email address.</p>
<p>Pydna can cache results from the following functions or methods:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#pydna.genbank.Genbank.nucleotide" title="pydna.genbank.Genbank.nucleotide"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.Genbank.nucleotide()</span></code></a>   Genbank_nucleotide</p></li>
<li><p><a class="reference internal" href="#pydna.amplify.Anneal" title="pydna.amplify.Anneal"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.Anneal()</span></code></a>               amplify_Anneal</p></li>
<li><p><a class="reference internal" href="#pydna.assembly.Assembly" title="pydna.assembly.Assembly"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.assembly.Assembly()</span></code></a>            assembly_Assembly</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.download.download_text()</span></code>       download.download_text</p></li>
<li><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.synced" title="pydna.dseqrecord.Dseqrecord.synced"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.synced()</span></code></a> Dseqrecord_synced</p></li>
</ul>
<p>These can be added separated by a comma to the cached_funcs entry
in <strong>pydna.ini</strong> file or the pydna_cached_funcs environment variable.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.open_log_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">open_log_folder</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_log_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.open_log_folder" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.get_env">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">get_env</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#get_env"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.get_env" title="Link to this definition"></a></dt>
<dd><p>Print a an ascii table containing all environmental variables.</p>
<p>Pydna related variables have names that starts with <cite>pydna_</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.logo">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">logo</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#logo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.logo" title="Link to this definition"></a></dt>
<dd><p>Ascii-art logotype of pydna.</p>
</dd></dl>

</section>
<section id="module-pydna.dseq">
<span id="pydna-dseq-module"></span><h2>pydna.dseq module<a class="headerlink" href="#module-pydna.dseq" title="Link to this heading"></a></h2>
<p>Provides the Dseq class for handling double stranded DNA sequences.</p>
<p>Dseq is a subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code>. The Dseq class
is mostly useful as a part of the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a> class
which can hold more meta data.</p>
<p>The Dseq class support the notion of circular and linear DNA topology.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.dseq.Dseq">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.dseq.</span></span><span class="sig-name descname"><span class="pre">Dseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></p>
<p>Dseq holds information for a double stranded DNA fragment.</p>
<p>Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>watson</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a string representing the watson (sense) DNA strand.</p></li>
<li><p><strong>crick</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – a string representing the crick (antisense) DNA strand.</p></li>
<li><p><strong>ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – A positive or negative number to describe the stagger between the
watson and crick strands.
see below for a detailed explanation.</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is linear, False that it is circular.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is circular, False that it is linear.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5’ end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a>).</p>
<p>There are three ways of creating a Dseq object directly listed below, but you can also
use the function Dseq.from_full_sequence_and_overhangs() to create a Dseq:</p>
<p>Only one argument (string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this. For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter is an integer describing the length of the
crick strand overhang in the 5’ end of the molecule.</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dsDNA</span>       <span class="n">overhang</span>

  <span class="n">nnn</span><span class="o">...</span>    <span class="mi">2</span>
<span class="n">nnnnn</span><span class="o">...</span>

 <span class="n">nnnn</span><span class="o">...</span>    <span class="mi">1</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>    <span class="mi">0</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">1</span>
 <span class="n">nnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">2</span>
  <span class="n">nnn</span><span class="o">...</span>
</pre></div>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>If the ovhg parameter is specified a crick strand also
needs to be supplied, otherwise an exception is raised.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
<span class="w">    </span><span class="k">else</span><span class="p">:</span>
<span class="gr">ValueError</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The shape of the fragment is set by circular = True, False</p>
<p>Note that both ends of the DNA fragment has to be compatible to set
circular = True.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s2">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">4</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ccccttt&quot;</span><span class="p">,</span><span class="s2">&quot;ccccaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-11)</span>
<span class="go">ccccttt</span>
<span class="go">    aaacccc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">-4</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>Coercing to string</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>A Dseq object can be longer that either the watson or crick strands.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>

<span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
      <span class="n">GATCCTTT</span>
<span class="n">AAAGCCCTA</span>
</pre></div>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this
will return the linearized sequence starting at X:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.trunc">
<span class="sig-name descname"><span class="pre">trunc</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">30</span></em><a class="headerlink" href="#pydna.dseq.Dseq.trunc" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.quick">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">quick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.quick"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.quick" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_string" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_representation">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dsdna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.from_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_representation" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_full_sequence_and_overhangs">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_full_sequence_and_overhangs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">full_sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">watson_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.from_full_sequence_and_overhangs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_full_sequence_and_overhangs" title="Link to this definition"></a></dt>
<dd><p>Create a linear Dseq object from a full sequence and the 3’ overhangs of each strand.</p>
<p>The order of the parameters is like this because the 3’ overhang of the crick strand is the one
on the left side of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>full_sequence</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The full sequence of the Dseq object.</p></li>
<li><p><strong>crick_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the crick strand in the 3’ end. Equivalent to Dseq.ovhg.</p></li>
<li><p><strong>watson_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the watson strand in the 5’ end.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Dseq object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AAAA</span>
<span class="go">TTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAAAA</span>
<span class="go">  TT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AA</span>
<span class="go">TTTTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAA</span>
<span class="go">  TTTT</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.mw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.mw" title="Link to this definition"></a></dt>
<dd><p>This method returns the molecular weight of the DNA molecule
in g/mol. The following formula is used:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MW</span> <span class="o">=</span> <span class="p">(</span><span class="n">A</span> <span class="n">x</span> <span class="mf">313.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">T</span> <span class="n">x</span> <span class="mf">304.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">C</span> <span class="n">x</span> <span class="mf">289.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">G</span> <span class="n">x</span> <span class="mf">329.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">N</span> <span class="n">x</span> <span class="mf">308.9</span><span class="p">)</span> <span class="o">+</span> <span class="mf">79.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.upper" title="Link to this definition"></a></dt>
<dd><p>Return an upper case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.lower" title="pydna.dseq.Dseq.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.lower" title="Link to this definition"></a></dt>
<dd><p>Return a lower case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.upper" title="pydna.dseq.Dseq.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sub</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_SeqAbstractBaseClass</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_sys.maxsize</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.find" title="Link to this definition"></a></dt>
<dd><p>This method behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<p>Returns -1 if the subsequence is NOT found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sub</strong> (<em>string</em><em> or </em><em>Seq object</em>) – a string or another Seq object to look for.</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice start.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice end.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;atcgactgacgtgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">atcgactgacgtgtt</span>
<span class="go">tagctgactgcacaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;gac&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;taa&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.dseq.Dseq.rc" title="Link to this definition"></a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.shifted" title="Link to this definition"></a></dt>
<dd><p>Shifted version of a circular Dseq object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.looped" title="Link to this definition"></a></dt>
<dd><p>Circularized Dseq object.</p>
<p>This can only be done if the two ends are compatible,
otherwise a TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
<span class="w">    </span><span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="x">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.tolinear" title="Link to this definition"></a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[:]</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.five_prime_end">
<span class="sig-name descname"><span class="pre">five_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.five_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.five_prime_end" title="Link to this definition"></a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.three_prime_end" title="pydna.dseq.Dseq.three_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.three_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.three_prime_end">
<span class="sig-name descname"><span class="pre">three_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.three_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.three_prime_end" title="Link to this definition"></a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.five_prime_end" title="pydna.dseq.Dseq.five_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.five_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.watson_ovhg">
<span class="sig-name descname"><span class="pre">watson_ovhg</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.watson_ovhg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.watson_ovhg" title="Link to this definition"></a></dt>
<dd><p>Returns the overhang of the watson strand at the three prime.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.fill_in">
<span class="sig-name descname"><span class="pre">fill_in</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.fill_in"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.fill_in" title="Link to this definition"></a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow <a class="footnote-reference brackets" href="#id2" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment">http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.transcribe">
<span class="sig-name descname"><span class="pre">transcribe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Seq</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.transcribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.transcribe" title="Link to this definition"></a></dt>
<dd><p>Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.</p>
<p>Following the usual convention, the sequence is interpreted as the
coding strand of the DNA double helix, not the template strand. This
means we can get the RNA sequence just by switching T to U.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">transcribe</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
<p>Trying to transcribe an RNA sequence has no effect.
If you have a nucleotide sequence which might be DNA or RNA
(or even a mixture), calling the transcribe method will ensure
any T becomes U.</p>
<p>Trying to transcribe a protein sequence will replace any
T for Threonine with U for Selenocysteine, which has no
biologically plausible rational.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Seq</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.translate" title="Link to this definition"></a></dt>
<dd><p>Translate..</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.mung">
<span class="sig-name descname"><span class="pre">mung</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.mung"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.mung" title="Link to this definition"></a></dt>
<dd><p>Simulates treatment a nuclease with 5’-3’ and 3’-5’ single
strand specific exonuclease activity (such as mung bean nuclease <a class="footnote-reference brackets" href="#id4" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">ggatcc</span>    <span class="o">-&gt;</span>     <span class="n">gatcc</span>
     <span class="n">ctaggg</span>          <span class="n">ctagg</span>

     <span class="n">ggatcc</span>   <span class="o">-&gt;</span>      <span class="n">ggatc</span>
    <span class="n">tcctag</span>            <span class="n">cctag</span>

<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
<span class="n">caaa</span>
 <span class="n">tttc</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
 <span class="n">aaac</span>
<span class="n">gttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Mung_bean_nuclease">http://en.wikipedia.org/wiki/Mung_bean_nuclease</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.T4">
<span class="sig-name descname"><span class="pre">T4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.T4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.T4" title="Link to this definition"></a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.t4">
<span class="sig-name descname"><span class="pre">t4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.dseq.Dseq.t4" title="Link to this definition"></a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.exo1_front">
<span class="sig-name descname"><span class="pre">exo1_front</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.exo1_front"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.exo1_front" title="Link to this definition"></a></dt>
<dd><p>5’-3’ resection at the start (left side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.exo1_end">
<span class="sig-name descname"><span class="pre">exo1_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.exo1_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.exo1_end" title="Link to this definition"></a></dt>
<dd><p>5’-3’ resection at the end (right side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.no_cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch not cutting sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.unique_cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="headerlink" href="#pydna.dseq.Dseq.once_cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.twice_cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence twice.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.n_cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting n times.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cutters" title="Link to this definition"></a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence at least once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.seguid" title="Link to this definition"></a></dt>
<dd><p>SEGUID checksum for the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.isblunt">
<span class="sig-name descname"><span class="pre">isblunt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.isblunt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.isblunt" title="Link to this definition"></a></dt>
<dd><p>isblunt.</p>
<p>Return True if Dseq is linear and blunt and
false if staggered or circular.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;atcg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> gat</span>
<span class="go">gcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;gatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">gat</span>
<span class="go">ctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#slice" title="(in Python v3.8)"><span class="pre">slice</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cas9" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.terminal_transferase" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><span class="pre">DseqType</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cut" title="Link to this definition"></a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there are no cuts, an empty list is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>enzyme object</em><em> or </em><em>iterable</em><em> of </em><em>such objects</em>) – A Bio.Restriction.XXX restriction objects or iterable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>frags</strong> – list of Dseq objects formed by the digestion</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;class &#39;tuple&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">frag</span><span class="p">))</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cutsite_is_valid">
<span class="sig-name descname"><span class="pre">cutsite_is_valid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cutsite_is_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cutsite_is_valid" title="Link to this definition"></a></dt>
<dd><p>Returns False if:
- Cut positions fall outside the sequence (could be moved to Biopython)
- Overhang is not double stranded
- Recognition site is not double stranded or is outside the sequence
- For enzymes that cut twice, it checks that at least one possibility is valid</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cutsites">
<span class="sig-name descname"><span class="pre">get_cutsites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.get_cutsites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cutsites" title="Link to this definition"></a></dt>
<dd><p>Returns a list of cutsites, represented represented as <cite>((cut_watson, ovhg), enz)</cite>:</p>
<ul class="simple">
<li><p><cite>cut_watson</cite> is a positive integer contained in <cite>[0,len(seq))</cite>, where <cite>seq</cite> is the sequence
that will be cut. It represents the position of the cut on the watson strand, using the full
sequence as a reference. By “full sequence” I mean the one you would get from <cite>str(Dseq)</cite>.</p></li>
<li><p><cite>ovhg</cite> is the overhang left after the cut. It has the same meaning as <cite>ovhg</cite> in
the <cite>Bio.Restriction</cite> enzyme objects, or pydna’s <cite>Dseq</cite> property.</p></li>
<li><dl class="simple">
<dt><cite>enz</cite> is the enzyme object. It’s not necessary to perform the cut, but can be</dt><dd><p>used to keep track of which enzyme was used.</p>
</dd>
</dl>
</li>
</ul>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>Union</em><em>[</em><em>_RestrictionBatch</em><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;AAGAATTCAAGAATTC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">)</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
</pre></div>
</div>
<p><cite>cut_watson</cite> is defined with respect to the “full sequence”, not the
watson strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaa&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span>
<span class="go">Dseq(-10)</span>
<span class="go"> aGAATTCaa</span>
<span class="go">ttCTTAAGtt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((3, -4), EcoRI)]</span>
</pre></div>
</div>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((1, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.left_end_position">
<span class="sig-name descname"><span class="pre">left_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.left_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.left_end_position" title="Link to this definition"></a></dt>
<dd><p>The index in the full sequence of the watson and crick start positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AAA</span>              <span class="n">AA</span>               <span class="n">AAT</span>
 <span class="n">TT</span>             <span class="n">TTT</span>               <span class="n">TTT</span>
<span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="n">Returns</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>    <span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.right_end_position">
<span class="sig-name descname"><span class="pre">right_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.right_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.right_end_position" title="Link to this definition"></a></dt>
<dd><p>The index in the full sequence of the watson and crick end positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<p><code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">AAA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AAA</span>
<span class="pre">TT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>
<span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">2)</span>&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(2,</span> <span class="pre">3)</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">3)</span>
<span class="pre">`</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cut_parameters">
<span class="sig-name descname"><span class="pre">get_cut_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_left</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.get_cut_parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cut_parameters" title="Link to this definition"></a></dt>
<dd><p>For a given cut expressed as ((cut_watson, ovhg), enz), returns
a tuple (cut_watson, cut_crick, ovhg).</p>
<ul class="simple">
<li><p>cut_watson: see get_cutsites docs</p></li>
<li><p>cut_crick: equivalent of cut_watson in the crick strand</p></li>
<li><p>ovhg: see get_cutsites docs</p></li>
</ul>
<p>The cut can be None if it represents the left or right end of the sequence.
Then it will return the position of the watson and crick ends with respect
to the “full sequence”. The <cite>is_left</cite> parameter is only used in this case.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.apply_cut" title="Link to this definition"></a></dt>
<dd><p>Extracts a subfragment of the sequence between two cuts.</p>
<p>For more detail see the documentation of get_cutsite_pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
<li><p><strong>right_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">p3</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span>
<span class="go">(None, ((3, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p1</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">aaG</span>
<span class="go">ttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span>
<span class="go">(((3, -4), EcoRI), ((11, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p2</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p3</span>
<span class="go">(((11, -4), EcoRI), None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p3</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">AATTCaa</span>
<span class="go">    Gtt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span>
<span class="go">(((6, -4), EcoRI), ((6, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">pair</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cutsite_pairs">
<span class="sig-name descname"><span class="pre">get_cutsite_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.get_cutsite_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cutsite_pairs" title="Link to this definition"></a></dt>
<dd><p>Returns pairs of cutsites that render the edges of the resulting fragments.</p>
<p>A fragment produced by restriction is represented by a tuple of length 2 that
may contain cutsites or <cite>None</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><p>Two cutsites: represents the extraction of a fragment between those two
cutsites, in that orientation. To represent the opening of a circular
molecule with a single cutsite, we put the same cutsite twice.</p></li>
<li><p><cite>None</cite>, cutsite: represents the extraction of a fragment between the left
edge of linear sequence and the cutsite.</p></li>
<li><p>cutsite, <cite>None</cite>: represents the extraction of a fragment between the cutsite
and the right edge of a linear sequence.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cutsites</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None],<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(None, ((3, -4), EcoRI)), (((3, -4), EcoRI), ((11, -4), EcoRI)), (((11, -4), EcoRI), None)]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(((6, -4), EcoRI), ((6, -4), EcoRI))]</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.dseqrecord">
<span id="pydna-dseqrecord-module"></span><h2>pydna.dseqrecord module<a class="headerlink" href="#module-pydna.dseqrecord" title="Link to this heading"></a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a> class, for handling double stranded
DNA sequences. The Dseqrecord holds sequence information in the form of a <a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseq.Dseq</span></code></a>
object. The Dseq and Dseqrecord classes are subclasses of Biopythons
Seq and SeqRecord classes, respectively.</p>
<p>The Dseq and Dseqrecord classes support the notion of circular and linear DNA topology.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.dseqrecord.</span></span><span class="sig-name descname"><span class="pre">Dseqrecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord <a class="footnote-reference brackets" href="#id6" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a> class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.</p>
<p>The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases.</p>
<p>Dseqrecord objects can be read or parsed from sequences in FASTA, EMBL or Genbank formats.
See the <a class="reference internal" href="#module-pydna.readers" title="pydna.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.readers</span></code></a> and <a class="reference internal" href="#module-pydna.parsers" title="pydna.parsers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.parsers</span></code></a> modules for further information.</p>
<p>There is a short representation associated with the Dseqrecord.
<code class="docutils literal notranslate"><span class="pre">Dseqrecord(-3)</span></code> represents a linear sequence of length 2
while <code class="docutils literal notranslate"><span class="pre">Dseqrecord(o7)</span></code>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length. This length can be larger
than each strand alone if they are staggered as in the example below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<em>string</em><em>, </em><em>Seq</em><em>, </em><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>, </em><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><em>Dseq</em></a><em> or </em><em>other Dseqrecord object</em>) – This data will be used to form the seq property</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">3</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_string" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.from_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.circular">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">circular</span></span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.circular" title="Link to this definition"></a></dt>
<dd><p>The circular property can not be set directly.
Use <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.looped" title="pydna.dseqrecord.Dseqrecord.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.m">
<span class="sig-name descname"><span class="pre">m</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.m"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.m" title="Link to this definition"></a></dt>
<dd><p>This method returns the mass of the DNA molecule in grams. This is
calculated as the product between the molecular weight of the Dseq object
and the</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.extract_feature" title="Link to this definition"></a></dt>
<dd><p>Extracts a feature and creates a new Dseqrecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates the feature to extract</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-2)</span>
<span class="go">gt</span>
<span class="go">ca</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.add_feature" title="Link to this definition"></a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4), strand=1), type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.seguid" title="Link to this definition"></a></dt>
<dd><p>Url safe SEGUID for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be part of a URL.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;ldseguid=TEwydy0ugvGXh3VJnVwgtxoyDQA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.looped" title="Link to this definition"></a></dt>
<dd><p>Circular version of the Dseqrecord object.</p>
<p>The underlying linear Dseq object has to have compatible ends.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.looped" title="pydna.dseq.Dseq.looped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.looped</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.tolinear" title="Link to this definition"></a></dt>
<dd><p>Returns a linear, blunt copy of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.terminal_transferase" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.format">
<span class="sig-name descname"><span class="pre">format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.format" title="Link to this definition"></a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO <a class="footnote-reference brackets" href="#id8" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>. Default is “gb” which is short for Genbank.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;gb&quot;</span><span class="p">))</span>
<span class="go">LOCUS       name                       3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  description.</span>
<span class="go">ACCESSION   id</span>
<span class="go">VERSION     id</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id8" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">4</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.write" title="Link to this definition"></a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO for writing <a class="footnote-reference brackets" href="#id10" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>. Default
is “gb” which is short for Genbank. Note that Biopython SeqIO reads
more formats than it writes.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format.</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;.&lt;f&gt;</span></code></p>
<p>Where &lt;f&gt; is “gb” by default. If the filename already exists and
AND the sequence it contains is different, a new file name will be
used so that the old file is not lost:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;_NEW.&lt;f&gt;</span></code></p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id10" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">5</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find_aminoacids">
<span class="sig-name descname"><span class="pre">find_aminoacids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.find_aminoacids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids" title="Link to this definition"></a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find_aa">
<span class="sig-name descname"><span class="pre">find_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aa" title="Link to this definition"></a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.map_trace_files">
<span class="sig-name descname"><span class="pre">map_trace_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.map_trace_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.map_trace_files" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.linearize">
<span class="sig-name descname"><span class="pre">linearize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.linearize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.linearize" title="Link to this definition"></a></dt>
<dd><p>Similar to <cite>:func:cut</cite>.</p>
<p>Throws an exception if there is not excactly one cut
i.e. none or more than one digestion products.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.no_cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.unique_cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.once_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.once_cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.twice_cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.n_cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cutters" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.number_of_cuts">
<span class="sig-name descname"><span class="pre">number_of_cuts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.number_of_cuts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts" title="Link to this definition"></a></dt>
<dd><p>The number of cuts by digestion with the Restriction enzymes
contained in the iterable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cas9" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.synced">
<span class="sig-name descname"><span class="pre">synced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.synced"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.synced" title="Link to this definition"></a></dt>
<dd><p>This method returns a new circular sequence (Dseqrecord object), which has been rotated
in such a way that there is maximum overlap between the sequence and
ref, which may be a string, Biopython Seq, SeqRecord object or
another Dseqrecord object.</p>
<p>The reason for using this could be to rotate a new recombinant plasmid so
that it starts at the same position after cloning. See the example below:</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;gaat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">gaat</span>
<span class="go">ctta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">a</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-4)</span>
<span class="go">atga</span>
<span class="go">tact</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span> <span class="o">=</span> <span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="n">insert</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">atgaCCC</span>
<span class="go">tactGGG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">gaCCCat</span>
<span class="go">ctGGGta</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.upper" title="Link to this definition"></a></dt>
<dd><p>Returns an uppercase copy.
&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord
&gt;&gt;&gt; my_seq = Dseqrecord(“aAa”)
&gt;&gt;&gt; my_seq.seq
Dseq(-3)
aAa
tTt
&gt;&gt;&gt; upper = my_seq.upper()
&gt;&gt;&gt; upper.seq
Dseq(-3)
AAA
TTT
&gt;&gt;&gt;</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.lower" title="pydna.dseqrecord.Dseqrecord.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.lower" title="Link to this definition"></a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.upper" title="pydna.dseqrecord.Dseqrecord.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.orfs">
<span class="sig-name descname"><span class="pre">orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.orfs" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.orfs_to_features">
<span class="sig-name descname"><span class="pre">orfs_to_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.orfs_to_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.orfs_to_features" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_gb_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.copy_gb_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_fasta_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.copy_fasta_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[48;5;11m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[0m'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.figure" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.shifted" title="Link to this definition"></a></dt>
<dd><p>Circular Dseqrecord with a new origin &lt;shift&gt;.</p>
<p>This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">GAAAT</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">CTTTA</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></code></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">shift</span></code> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin at the | symbol:</div>
<div class="line"><br /></div>
</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">G|AAAT</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">C|TTTA</span></code></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">AAATG</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">TTTAC</span></code></div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cut" title="Link to this definition"></a></dt>
<dd><p>Digest a Dseqrecord object with one or more restriction enzymes.</p>
<p>returns a list of linear Dseqrecords. If there are no cuts, an empty
list is returned.</p>
<p>See also <code class="xref py py-func docutils literal notranslate"><span class="pre">Dseq.cut()</span></code>
:param enzymes: A Bio.Restriction.XXX restriction object or iterable of such.
:type enzymes: enzyme object or iterable of such objects</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>Dseqrecord_frags</strong> – list of Dseqrecord objects formed by the digestion</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">(Dseqrecord(-5), Dseqrecord(-5))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.apply_cut" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pydna.amplicon">
<span id="pydna-amplicon-module"></span><h2>pydna.amplicon module<a class="headerlink" href="#module-pydna.amplicon" title="Link to this heading"></a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code></a> class for PCR simulation.
This class is not meant to be use directly but is
used by the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.amplicon.</span></span><span class="sig-name descname"><span class="pre">Amplicon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<p>The Amplicon class holds information about a PCR reaction involving two
primers and one template. This class is used by the Anneal class and is not
meant to be instantiated directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_primer</strong> (<a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>(</em><em>Biopython</em><em>)</em>) – SeqRecord object holding the forward (sense) primer</p></li>
<li><p><strong>reverse_primer</strong> (<a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>(</em><em>Biopython</em><em>)</em>) – SeqRecord object holding the reverse (antisense) primer</p></li>
<li><p><strong>template</strong> (<a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><em>Dseqrecord</em></a>) – Dseqrecord object holding the template (circular or linear)</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.from_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.from_SeqRecord" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplicon.Amplicon.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.figure" title="Link to this definition"></a></dt>
<dd><p>This method returns a simple figure of the two primers binding
to a part of the template.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">5</span><span class="n">tacactcaccgtctatcattatc</span><span class="o">...</span><span class="n">cgactgtatcatctgatagcac3</span>
                           <span class="o">||||||||||||||||||||||</span>
                          <span class="mi">3</span><span class="n">gctgacatagtagactatcgtg5</span>
<span class="mi">5</span><span class="n">tacactcaccgtctatcattatc3</span>
 <span class="o">|||||||||||||||||||||||</span>
<span class="mi">3</span><span class="n">atgtgagtggcagatagtaatag</span><span class="o">...</span><span class="n">gctgacatagtagactatcgtg5</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>figure</strong> – A string containing a text representation of the primers
annealing on the template (see example above).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.set_forward_primer_footprint">
<span class="sig-name descname"><span class="pre">set_forward_primer_footprint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.set_forward_primer_footprint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.set_forward_primer_footprint" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.set_reverse_primer_footprint">
<span class="sig-name descname"><span class="pre">set_reverse_primer_footprint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.set_reverse_primer_footprint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.set_reverse_primer_footprint" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.program">
<span class="sig-name descname"><span class="pre">program</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.program" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.dbd_program">
<span class="sig-name descname"><span class="pre">dbd_program</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.dbd_program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.dbd_program" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.primers">
<span class="sig-name descname"><span class="pre">primers</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.primers" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pydna.amplify">
<span id="pydna-amplify-module"></span><h2>pydna.amplify module<a class="headerlink" href="#module-pydna.amplify" title="Link to this heading"></a></h2>
<p>This module provide the <a class="reference internal" href="#pydna.amplify.Anneal" title="pydna.amplify.Anneal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Anneal</span></code></a> class and the <a class="reference internal" href="#pydna.amplify.pcr" title="pydna.amplify.pcr"><code class="xref py py-func docutils literal notranslate"><span class="pre">pcr()</span></code></a> function
for PCR simulation. The pcr function is simpler to use, but expects only one
PCR product. The Anneal class should be used if more flexibility is required.</p>
<p>Primers with 5’ tails as well as inverse PCR on circular templates are handled
correctly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.amplify.Anneal">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.amplify.</span></span><span class="sig-name descname"><span class="pre">Anneal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplify.html#Anneal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplify.Anneal" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Anneal class has the following important attributes:</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.forward_primers">
<span class="sig-name descname"><span class="pre">forward_primers</span></span><a class="headerlink" href="#pydna.amplify.Anneal.forward_primers" title="Link to this definition"></a></dt>
<dd><p>Description of <cite>forward_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.reverse_primers">
<span class="sig-name descname"><span class="pre">reverse_primers</span></span><a class="headerlink" href="#pydna.amplify.Anneal.reverse_primers" title="Link to this definition"></a></dt>
<dd><p>Description of <cite>reverse_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.template">
<span class="sig-name descname"><span class="pre">template</span></span><a class="headerlink" href="#pydna.amplify.Anneal.template" title="Link to this definition"></a></dt>
<dd><p>A copy of the template argument. Primers annealing sites has been
added as features that can be visualized in a seqence editor such as
ApE.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.limit">
<span class="sig-name descname"><span class="pre">limit</span></span><a class="headerlink" href="#pydna.amplify.Anneal.limit" title="Link to this definition"></a></dt>
<dd><p>The limit of PCR primer annealing, default is 13 bp.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.products">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">products</span></span><a class="headerlink" href="#pydna.amplify.Anneal.products" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.report">
<span class="sig-name descname"><span class="pre">report</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplify.Anneal.report" title="Link to this definition"></a></dt>
<dd><p>returns a short report describing if or where primer
anneal on the template.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.amplify.pcr">
<span class="sig-prename descclassname"><span class="pre">pydna.amplify.</span></span><span class="sig-name descname"><span class="pre">pcr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplify.html#pcr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplify.pcr" title="Link to this definition"></a></dt>
<dd><p>pcr is a convenience function for the Anneal class to simplify its
usage, especially from the command line. If more than one or no PCR
product is formed, a ValueError is raised.</p>
<p>args is any iterable of Dseqrecords or an iterable of iterables of
Dseqrecords. args will be greedily flattened.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable containing sequence objects</em>) – Several arguments are also accepted.</p></li>
<li><p><strong>limit</strong> (<em>int = 13</em><em>, </em><em>optional</em>) – limit length of the annealing part of the primers.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>sequences in args could be of type:</p>
<ul class="simple">
<li><p>string</p></li>
<li><p>Seq</p></li>
<li><p>SeqRecord (or subclass)</p></li>
<li><p>Dseqrecord (or sublcass)</p></li>
</ul>
<p>The last sequence will be assumed to be the template while
all preceeding sequences will be assumed to be primers.</p>
<p>This is a powerful function, use with care!</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>product</strong> – An <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> object representing the PCR
product. The direction of the PCR product will be the same as for
the template sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.readers</span> <span class="kn">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.primer</span> <span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;cgactgtatcatctgatagcac&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.assembly">
<span id="pydna-assembly-module"></span><h2>pydna.assembly module<a class="headerlink" href="#module-pydna.assembly" title="Link to this heading"></a></h2>
<p>Assembly of sequences by homologous recombination.</p>
<p>Should also be useful for related techniques such as Gibson assembly and fusion
PCR. Given a list of sequences (Dseqrecords), all sequences are analyzed for
shared homology longer than the set limit.</p>
<p>A graph is constructed where each overlapping region form a node and
sequences separating the overlapping regions form edges.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>            <span class="o">--</span> <span class="n">A</span> <span class="o">--</span>
<span class="n">catgatctacgtatcgtgt</span>     <span class="o">--</span> <span class="n">B</span> <span class="o">--</span>
            <span class="n">atcgtgtactgtcatattc</span>
                        <span class="n">catattcaaagttct</span>



<span class="o">--</span><span class="n">x</span><span class="o">--&gt;</span> <span class="n">A</span> <span class="o">--</span><span class="n">y</span><span class="o">--&gt;</span> <span class="n">B</span> <span class="o">--</span><span class="n">z</span><span class="o">--&gt;</span>   <span class="p">(</span><span class="n">Graph</span><span class="p">)</span>

<span class="n">Nodes</span><span class="p">:</span>

<span class="n">A</span> <span class="p">:</span> <span class="n">atcgtgt</span>
<span class="n">B</span> <span class="p">:</span> <span class="n">catattc</span>

<span class="n">Edges</span><span class="p">:</span>

<span class="n">x</span> <span class="p">:</span> <span class="n">catgatctacgt</span>
<span class="n">y</span> <span class="p">:</span> <span class="n">actgt</span>
<span class="n">z</span> <span class="p">:</span> <span class="n">aaagttct</span>
</pre></div>
</div>
<p>The NetworkX package is used to trace linear and circular paths through the
graph.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.assembly.Assembly">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.assembly.</span></span><span class="sig-name descname"><span class="pre">Assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">common_sub_strings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/assembly.html#Assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly.Assembly" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Assembly of a list of linear DNA fragments into linear or circular
constructs. The Assembly is meant to replace the Assembly method as it
is easier to use. Accepts a list of Dseqrecords (source fragments) to
initiate an Assembly object. Several methods are available for analysis
of overlapping sequences, graph construction and assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fragments</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – a list of Dseqrecord objects.</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – The shortest shared homology to be considered</p></li>
<li><p><strong>algorithm</strong> (<em>function</em><em>, </em><em>optional</em>) – The algorithm used to determine the shared sequences.</p></li>
<li><p><strong>max_nodes</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The maximum number of nodes in the graph. This can be tweaked to
manage sequences with a high number of shared sub sequences.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">((</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">),</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 33bp 34bp 35bp</span>
<span class="go">limit(bp)....: 14</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>
<span class="go">[Contig(o59), Contig(o59)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">watson</span>
<span class="go">&#39;acgatgctatactgCCCCCtgtgctgtgctctaTTTTTtattctggctgtatcGGGGGt&#39;</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly.Assembly.assemble_linear">
<span class="sig-name descname"><span class="pre">assemble_linear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.assembly.Assembly.assemble_linear" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly.Assembly.assemble_circular">
<span class="sig-name descname"><span class="pre">assemble_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.assembly.Assembly.assemble_circular" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pydna.common_sub_strings">
<span id="pydna-common-sub-strings-module"></span><h2>pydna.common_sub_strings module<a class="headerlink" href="#module-pydna.common_sub_strings" title="Link to this heading"></a></h2>
<p>This module is based on the Py-rstr-max package that
was written by Romain Brixtel (rbrixtel_at_gmail_dot_com)
(<a class="reference external" href="https://brixtel.users.greyc.fr">https://brixtel.users.greyc.fr</a>) and is available from
<a class="reference external" href="https://code.google.com/p/py-rstr-max">https://code.google.com/p/py-rstr-max</a>
<a class="reference external" href="https://github.com/gip0/py-rstr-max">https://github.com/gip0/py-rstr-max</a>
the original code was covered by an MIT licence.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.common_sub_strings.common_sub_strings">
<span class="sig-prename descclassname"><span class="pre">pydna.common_sub_strings.</span></span><span class="sig-name descname"><span class="pre">common_sub_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stringx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">stringy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#common_sub_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.common_sub_strings.common_sub_strings" title="Link to this definition"></a></dt>
<dd><p>Finds all common substrings between stringx and stringy, and returns
them sorted by length.</p>
<p>This function is case sensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stringx</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>stringy</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>[(startx1, starty1, length1),(startx2, starty2, length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a> of <a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.common_sub_strings.terminal_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.common_sub_strings.</span></span><span class="sig-name descname"><span class="pre">terminal_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stringx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">stringy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#terminal_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.common_sub_strings.terminal_overlap" title="Link to this definition"></a></dt>
<dd><p>Finds the the flanking common substrings between stringx and stringy
longer than limit. This means that the results only contains substrings
that starts or ends at the the ends of stringx and stringy.</p>
<p>This function is case sensitive.</p>
<p>returns a list of tuples describing the substrings
The list is sorted longest -&gt; shortest.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stringx</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>stringy</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>[(startx1,starty1,length1),(startx2,starty2,length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a> of <a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.common_sub_strings</span> <span class="kn">import</span> <span class="n">terminal_overlap</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">terminal_overlap</span><span class="p">(</span><span class="s2">&quot;agctatgtatcttgcatcgta&quot;</span><span class="p">,</span> <span class="s2">&quot;gcatcgtagtctatttgcttac&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="go">[(13, 0, 8)]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>             <span class="o">&lt;--</span> <span class="mi">8</span> <span class="o">-&gt;</span>
<span class="o">&lt;----</span> <span class="mi">13</span> <span class="o">---&gt;</span>
<span class="n">agctatgtatcttgcatcgta</span>                    <span class="n">stringx</span>
             <span class="n">gcatcgtagtctatttgcttac</span>      <span class="n">stringy</span>
             <span class="mi">0</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.contig">
<span id="pydna-contig-module"></span><h2>pydna.contig module<a class="headerlink" href="#module-pydna.contig" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.contig.Contig">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.contig.</span></span><span class="sig-name descname"><span class="pre">Contig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<p>This class holds information about a DNA assembly. This class is instantiated by
the <code class="xref py py-class docutils literal notranslate"><span class="pre">Assembly</span></code> class and is not meant to be used directly.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.from_string" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.from_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.from_SeqRecord" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.contig.Contig.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.detailed_figure">
<span class="sig-name descname"><span class="pre">detailed_figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.detailed_figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.detailed_figure" title="Link to this definition"></a></dt>
<dd><p>Returns a text representation of the assembled fragments.</p>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">||||||||||||||</span>
<span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
                                      <span class="n">TATTCTGGCTGTATC</span>
                                      <span class="n">tattctggctgtatcGGGGGtacgatgctatactg</span>
                                                           <span class="n">ACGATGCTATACTG</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.figure" title="Link to this definition"></a></dt>
<dd><p>Compact ascii representation of the assembled fragments.</p>
<p>Each fragment is represented by:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Size</span> <span class="n">of</span> <span class="n">common</span> <span class="mi">5</span><span class="s1">&#39; substring|Name and size of DNA fragment|</span>
<span class="n">Size</span> <span class="n">of</span> <span class="n">common</span> <span class="mi">5</span><span class="s1">&#39; substring</span>
</pre></div>
</div>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">frag20</span><span class="o">|</span> <span class="mi">6</span>
       \\<span class="o">/</span>
       <span class="o">/</span>\\
        <span class="mi">6</span><span class="o">|</span><span class="n">frag23</span><span class="o">|</span> <span class="mi">6</span>
                 \\<span class="o">/</span>
                 <span class="o">/</span>\\
                  <span class="mi">6</span><span class="o">|</span><span class="n">frag14</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="o">-|</span><span class="mi">2577</span><span class="o">|</span><span class="mi">61</span>
<span class="o">|</span>       \\<span class="o">/</span>
<span class="o">|</span>       <span class="o">/</span>\\
<span class="o">|</span>       <span class="mi">61</span><span class="o">|</span><span class="mi">5681</span><span class="o">|</span><span class="mi">98</span>
<span class="o">|</span>               \\<span class="o">/</span>
<span class="o">|</span>               <span class="o">/</span>\\
<span class="o">|</span>               <span class="mi">98</span><span class="o">|</span><span class="mi">2389</span><span class="o">|</span><span class="mi">557</span>
<span class="o">|</span>                       \\<span class="o">/</span>
<span class="o">|</span>                       <span class="o">/</span>\\
<span class="o">|</span>                       <span class="mi">557</span><span class="o">-</span>
<span class="o">|</span>                          <span class="o">|</span>
 <span class="o">--------------------------</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.design">
<span id="pydna-design-module"></span><h2>pydna.design module<a class="headerlink" href="#module-pydna.design" title="Link to this heading"></a></h2>
<p>This module contain functions for primer design for various purposes.</p>
<ul class="simple">
<li><p>:func:primer_design for designing primers for a sequence or a matching primer for an existing primer. Returns an <code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code> object (same as the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module returns).</p></li>
<li><p>:func:assembly_fragments Adds tails to primers for a linear assembly through homologous recombination or Gibson assembly.</p></li>
<li><p>:func:circular_assembly_fragments Adds tails to primers for a circular assembly through homologous recombination or Gibson assembly.</p></li>
</ul>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.primer_design">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">primer_design</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">55.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimate_function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#primer_design"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.primer_design" title="Link to this definition"></a></dt>
<dd><p>This function designs a forward primer and a reverse primer for PCR amplification
of a given template sequence.</p>
<p>The template argument is a Dseqrecord object or equivalent containing the template sequence.</p>
<p>The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).
One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).</p>
<p>The limit argument is the minimum length of the primer. The default value is 13.</p>
<p>If one of the primers is given, the other primer is designed to match in terms of Tm.
If both primers are designed, they will be designed to target_tm</p>
<p>tm_func is a function that takes an ascii string representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be substituted for a custom made function.</p>
<p>estimate_function is a tm_func-like function that is used to get a first guess for the primer design, that is then used as starting
point for the final result. This is useful when the tm_func function is slow to calculate (e.g. it relies on an
external API, such as the NEB primer design API). The estimate_function should be faster than the tm_func function.
The default value is <cite>None</cite>.
To use the default <cite>tm_func</cite> as estimate function to get the NEB Tm faster, you can do:
<cite>primer_design(dseqr, target_tm=55, tm_func=tm_neb, estimate_function=tm_default)</cite>.</p>
<p>The function returns a pydna.amplicon.Amplicon class instance. This object has
the object.forward_primer and object.reverse_primer properties which contain the designed primers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><em>pydna.dseqrecord.Dseqrecord</em></a>) – a Dseqrecord object. The only required argument.</p></li>
<li><p><strong>fp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>rp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>target_tm</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a><em>, </em><em>optional</em>) – target tm for the primers, set to 55°C by default.</p></li>
<li><p><strong>tm_func</strong> (<em>function</em>) – Function used for tm calculation. This function takes an ascii string
representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be
substituted for a custom made function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>result</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">Dseqrecord(-64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span>
<span class="go">Amplicon(64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">f64 17-mer:5&#39;-atgactgctaacccttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 18-mer:5&#39;-catcgtaagtttcgaacg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ampl</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                     ||||||||||||||||||</span>
<span class="go">                    3gcaagctttgaatgctac5</span>
<span class="go">5atgactgctaacccttc3</span>
<span class="go"> |||||||||||||||||</span>
<span class="go">3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">f64 23-mer:5&#39;-GGATCCatgactgct..ttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">r64 24-mer:5&#39;-GGATCCcatcgtaag..acg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">      5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                           ||||||||||||||||||</span>
<span class="go">                          3gcaagctttgaatgctacCCTAGG5</span>
<span class="go">5GGATCCatgactgctaacccttc3</span>
<span class="go">       |||||||||||||||||</span>
<span class="go">      3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.primer</span> <span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttg&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myprimer&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp</span> <span class="o">=</span> <span class="n">pf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">myprimer 27-mer:5&#39;-atgactgctaaccct..ttg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 32-mer:5&#39;-catcgtaagtttcga..atc-3&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.assembly_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">assembly_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlink</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#assembly_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.assembly_fragments" title="Link to this definition"></a></dt>
<dd><p>This function return a list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects where
primers have been modified with tails so that the fragments can be fused in
the order they appear in the list by for example Gibson assembly or homologous
recombination.</p>
<p>Given that we have two linear <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects a and b</p>
<p>we can modify the reverse primer of a and forward primer of b with tails to allow
fusion by fusion PCR, Gibson assembly or in-vivo homologous recombination.
The basic requirements for the primers for the three techniques are the same.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>
<span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>
                  <span class="o">|||||</span>
                 <span class="o">&lt;</span><span class="n">gacgt</span>
<span class="n">agcct</span><span class="o">&gt;</span>
<span class="o">|||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                        <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
                       <span class="o">/</span>                     \
                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                                         <span class="o">|||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
                       <span class="n">TTTAT</span><span class="o">&gt;</span>
                       <span class="o">|||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

<span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
\<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>Design tailed primers incorporating a part of the next or previous fragment to be assembled.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATA</span>

<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>

                <span class="n">ctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by any of the techniques mentioned earlier to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p><strong>At least every second sequence object needs to be an Amplicon</strong></p>
<p>This rule exists because if a sequence object is that is not a PCR product
is to be fused with another fragment, that other fragment needs to be an Amplicon
so that the primer of the other object can be modified to include the whole stretch
of sequence homology needed for the fusion. See the example below where a is a
non-amplicon (a linear plasmid  vector for instance)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>           <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
<span class="o">/</span>                     \         <span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>   <span class="o">&lt;--&gt;</span>  <span class="n">TTTATATCGCATGACTCTTCTTT</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>                          <span class="o">&lt;</span><span class="n">AGAAA</span>
                                <span class="n">TTTAT</span><span class="o">&gt;</span>
                                <span class="o">|||||||||||||||||||||||</span>
                          <span class="o">&lt;--&gt;</span>  <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

     <span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
     <span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
     <span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
     \<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>In this case only the forward primer of b is fitted with a tail with a part a:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
         <span class="n">tcttggtctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by for example Gibson assembly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaacca</span>

                             <span class="n">TCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p>The overlap argument controls how many base pairs of overlap required between
adjacent sequence fragments. In the junction between Amplicons, tails with the
length of about half of this value is added to the two primers
closest to the junction.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;       &lt;
Amplicon1
         Amplicon2
         &gt;       &lt;

         ⇣

&gt;       &lt;-
Amplicon1
         Amplicon2
        -&gt;       &lt;
</pre></div>
</div>
<p>In the case of an Amplicon adjacent to a Dseqrecord object, the tail will
be twice as long (1*overlap) since the
recombining sequence is present entirely on this primer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Dseqrecd1
         Amplicon1
         &gt;       &lt;

         ⇣

Dseqrecd1
         Amplicon1
       --&gt;       &lt;
</pre></div>
</div>
<p>Note that if the sequence of DNA fragments starts or stops with an Amplicon,
the very first and very last prinmer will not be modified i.e. assembles are
always assumed to be linear. There are simple tricks around that for circular
assemblies depicted in the last two examples below.</p>
<p>The maxlink arguments controls the cut off length for sequences that will be
synhtesized by adding them to primers for the adjacent fragment(s). The
argument list may contain short spacers (such as spacers between fusion proteins).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Example 1: Linear assembly of PCR products (pydna.amplicon.Amplicon class objects) ------

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon4
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;-       -&gt;       &lt;-                      pydna.assembly.Assembly
Amplicon1         Amplicon3
         Amplicon2         Amplicon4     ➤  Amplicon1Amplicon2Amplicon3Amplicon4
        -&gt;       &lt;-       -&gt;       &lt;

Example 2: Linear assembly of alternating Amplicons and other fragments

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;--     --&gt;       &lt;--                     pydna.assembly.Assembly
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2     ➤  Amplicon1Dseqrecd1Amplicon2Dseqrecd2

Example 3: Linear assembly of alternating Amplicons and other fragments

Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2
         &gt;       &lt;       --&gt;       &lt;

                     ⇣
             pydna.design.assembly_fragments
                     ⇣
                                                  pydna.assembly.Assembly
Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2     ➤  Dseqrecd1Amplicon1Dseqrecd2Amplicon2
       --&gt;       &lt;--     --&gt;       &lt;

Example 4: Circular assembly of alternating Amplicons and other fragments

                 -&gt;       &lt;==
Dseqrecd1         Amplicon2
         Amplicon1         Dseqrecd1
       --&gt;       &lt;-
                     ⇣
                     pydna.design.assembly_fragments
                     ⇣
                                                   pydna.assembly.Assembly
                 -&gt;       &lt;==
Dseqrecd1         Amplicon2                    -Dseqrecd1Amplicon1Amplicon2-
         Amplicon1                       ➤    |                             |
       --&gt;       &lt;-                            -----------------------------

------ Example 5: Circular assembly of Amplicons

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;=       -&gt;       &lt;-
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
        -&gt;       &lt;-       +&gt;       &lt;

                     ⇣
             make new Amplicon using the Amplicon1.template and
             the last fwd primer and the first rev primer.
                     ⇣
                                                   pydna.assembly.Assembly
+&gt;       &lt;=       -&gt;       &lt;-
 Amplicon1         Amplicon3                  -Amplicon1Amplicon2Amplicon3-
          Amplicon2                      ➤   |                             |
         -&gt;       &lt;-                          -----------------------------
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects) – list Amplicon and Dseqrecord object for which fusion primers should be constructed.</p></li>
<li><p><strong>overlap</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Length of required overlap between fragments.</p></li>
<li><p><strong>maxlink</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Maximum length of spacer sequences that may be present in f. These will be included in tails for designed primers.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>seqs</strong> –</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Amplicon1</span><span class="p">,</span>
 <span class="n">Amplicon2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="kn">import</span> <span class="n">assembly_fragments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># We would like a circular recombination, so the first sequence has to be repeated</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa1</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">,</span><span class="n">fa2</span> <span class="o">=</span> <span class="n">assembly_fragments</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">a</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Since all fragments are Amplicons, we need to extract the rp of the 1st and fp of the last fragments.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">fa2</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">fa1</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">]</span>
<span class="go">[Amplicon(100), Amplicon(101), Amplicon(102)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fb</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fc</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;fa fb fc&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">([</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 100bp 101bp 102bp</span>
<span class="go">limit(bp)....: 25</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_linear</span><span class="p">()</span>
<span class="go">[Contig(-231), Contig(-166), Contig(-36)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go"> -|fa|36</span>
<span class="go">|     \/</span>
<span class="go">|     /\</span>
<span class="go">|     36|fb|36</span>
<span class="go">|           \/</span>
<span class="go">|           /\</span>
<span class="go">|           36|fc|36</span>
<span class="go">|                 \/</span>
<span class="go">|                 /\</span>
<span class="go">|                 36-</span>
<span class="go">|                    |</span>
<span class="go"> --------------------</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.circular_assembly_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">circular_assembly_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlink</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#circular_assembly_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.circular_assembly_fragments" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.download">
<span id="pydna-download-module"></span><h2>pydna.download module<a class="headerlink" href="#module-pydna.download" title="Link to this heading"></a></h2>
<p>Provides a function for downloading online text files.</p>
</section>
<section id="module-pydna.editor">
<span id="pydna-editor-module"></span><h2>pydna.editor module<a class="headerlink" href="#module-pydna.editor" title="Link to this heading"></a></h2>
<p>This module provides a class for opening a sequence using an editor
that accepts a file as a command line argument.</p>
<p>ApE - A plasmid Editor <a class="footnote-reference brackets" href="#id12" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a> is and excellent editor for this purpose.</p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id12" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">6</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biologylabs.utah.edu/jorgensen/wayned/ape/">http://biologylabs.utah.edu/jorgensen/wayned/ape/</a></p>
</aside>
</aside>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.editor.Editor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.editor.</span></span><span class="sig-name descname"><span class="pre">Editor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shell_command_for_editor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmpdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#Editor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.editor.Editor" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Editor class needs to be instantiated before use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shell_command_for_editor</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – String containing the path to the editor</p></li>
<li><p><strong>tmpdir</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – String containing path to the temprary directory where sequence
files are stored before opening.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#ape = pydna.Editor(&quot;tclsh8.6 /home/bjorn/.ApE/apeextractor/ApE.vfs/lib/app-AppMain/AppMain.tcl&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#ape.open(&quot;aaa&quot;) # This command opens the sequence in the ApE editor</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.editor.Editor.open">
<span class="sig-name descname"><span class="pre">open</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_to_open</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#Editor.open"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.editor.Editor.open" title="Link to this definition"></a></dt>
<dd><p>Open a sequence for editing in an external (DNA) editor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>args</strong> (<a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em> or </em><em>Dseqrecord object</em>)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.editor.ape">
<span class="sig-prename descclassname"><span class="pre">pydna.editor.</span></span><span class="sig-name descname"><span class="pre">ape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#ape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.editor.ape" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.gel">
<span id="pydna-gel-module"></span><h2>pydna.gel module<a class="headerlink" href="#module-pydna.gel" title="Link to this heading"></a></h2>
<p>docstring.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.gel.interpolator">
<span class="sig-prename descclassname"><span class="pre">pydna.gel.</span></span><span class="sig-name descname"><span class="pre">interpolator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mwstd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#interpolator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gel.interpolator" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.gel.gel">
<span class="sig-prename descclassname"><span class="pre">pydna.gel.</span></span><span class="sig-name descname"><span class="pre">gel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gel_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">margin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interpolator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">interpolator(mwstd=_mwstd)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#gel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gel.gel" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.genbank">
<span id="pydna-genbank-module"></span><h2>pydna.genbank module<a class="headerlink" href="#module-pydna.genbank" title="Link to this heading"></a></h2>
<p>This module provides a class for downloading sequences from genbank
called Genbank and an function that does the same thing called genbank.</p>
<p>The function can be used if the environmental variable <strong>pydna_email</strong> has
been set to a valid email address. The easiest way to do this permanantly is to edit the
<cite>pydna.ini</cite> file. See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbank.Genbank">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbank.</span></span><span class="sig-name descname"><span class="pre">Genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">users_email</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pydna'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbank.html#Genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbank.Genbank" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Class to facilitate download from genbank. It is easier and
quicker to use the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>users_email</strong> (<em>string</em>) – Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.genbank</span> <span class="kn">import</span> <span class="n">Genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gb</span><span class="o">=</span><span class="n">Genbank</span><span class="p">(</span><span class="s2">&quot;bjornjobb@gmail.com&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="s2">&quot;LP002422.1&quot;</span><span class="p">)</span>   <span class="c1"># &lt;- entry from genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbank.Genbank.nucleotide">
<span class="sig-name descname"><span class="pre">nucleotide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbank.Genbank.nucleotide" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbank.genbank">
<span class="sig-prename descclassname"><span class="pre">pydna.genbank.</span></span><span class="sig-name descname"><span class="pre">genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">accession</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CS570233.1'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbank.html#genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbank.genbank" title="Link to this definition"></a></dt>
<dd><p>Download a genbank nuclotide record.</p>
<p>This function takes the same paramenters as the
:func:pydna.genbank.Genbank.nucleotide method. The email address stored
in the <cite>pydna_email</cite> environment variable is used. The easiest way set
this permanantly is to edit the <cite>pydna.ini</cite> file.
See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
<p>if no accession is given, a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/CS570233.1">very short Genbank
entry</a>
is used as an example (see below). This can be useful for testing the
connection to Genbank.</p>
<p>Please note that this result is also cached by default by settings in
the pydna.ini file.
See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">LOCUS</span>       <span class="n">CS570233</span>                  <span class="mi">14</span> <span class="n">bp</span>    <span class="n">DNA</span>     <span class="n">linear</span>   <span class="n">PAT</span> <span class="mi">18</span><span class="o">-</span><span class="n">MAY</span><span class="o">-</span><span class="mi">2007</span>
<span class="n">DEFINITION</span>  <span class="n">Sequence</span> <span class="mi">6</span> <span class="kn">from</span> <span class="nn">Patent</span> <span class="n">WO2007025016</span><span class="o">.</span>
<span class="n">ACCESSION</span>   <span class="n">CS570233</span>
<span class="n">VERSION</span>     <span class="n">CS570233</span><span class="mf">.1</span>
<span class="n">KEYWORDS</span>    <span class="o">.</span>
<span class="n">SOURCE</span>      <span class="n">synthetic</span> <span class="n">construct</span>
  <span class="n">ORGANISM</span>  <span class="n">synthetic</span> <span class="n">construct</span>
            <span class="n">other</span> <span class="n">sequences</span><span class="p">;</span> <span class="n">artificial</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">REFERENCE</span>   <span class="mi">1</span>
  <span class="n">AUTHORS</span>   <span class="n">Shaw</span><span class="p">,</span><span class="n">R</span><span class="o">.</span><span class="n">W</span><span class="o">.</span> <span class="ow">and</span> <span class="n">Cottenoir</span><span class="p">,</span><span class="n">M</span><span class="o">.</span>
  <span class="n">TITLE</span>     <span class="n">Inhibition</span> <span class="n">of</span> <span class="n">metallo</span><span class="o">-</span><span class="n">beta</span><span class="o">-</span><span class="n">lactamase</span> <span class="n">by</span> <span class="n">double</span><span class="o">-</span><span class="n">stranded</span> <span class="n">dna</span>
  <span class="n">JOURNAL</span>   <span class="n">Patent</span><span class="p">:</span> <span class="n">WO</span> <span class="mi">2007025016</span><span class="o">-</span><span class="n">A1</span> <span class="mi">6</span> <span class="mi">01</span><span class="o">-</span><span class="n">MAR</span><span class="o">-</span><span class="mi">2007</span><span class="p">;</span>
            <span class="n">Texas</span> <span class="n">Tech</span> <span class="n">University</span> <span class="n">System</span> <span class="p">(</span><span class="n">US</span><span class="p">)</span>
<span class="n">FEATURES</span>             <span class="n">Location</span><span class="o">/</span><span class="n">Qualifiers</span>
     <span class="n">source</span>          <span class="mf">1..14</span>
                     <span class="o">/</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;synthetic construct&quot;</span>
                     <span class="o">/</span><span class="n">mol_type</span><span class="o">=</span><span class="s2">&quot;unassigned DNA&quot;</span>
                     <span class="o">/</span><span class="n">db_xref</span><span class="o">=</span><span class="s2">&quot;taxon:32630&quot;</span>
                     <span class="o">/</span><span class="n">note</span><span class="o">=</span><span class="s2">&quot;This is a 14bp aptamer inhibitor.&quot;</span>
<span class="n">ORIGIN</span>
        <span class="mi">1</span> <span class="n">atgttcctac</span> <span class="n">atga</span>
<span class="o">//</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.genbankfile">
<span id="pydna-genbankfile-module"></span><h2>pydna.genbankfile module<a class="headerlink" href="#module-pydna.genbankfile" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbankfile.</span></span><span class="sig-name descname"><span class="pre">GenbankFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.from_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.from_SeqRecord" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankfile.GenbankFile.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.genbankfixer">
<span id="pydna-genbankfixer-module"></span><h2>pydna.genbankfixer module<a class="headerlink" href="#module-pydna.genbankfixer" title="Link to this heading"></a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.genbankfixer.gbtext_clean" title="pydna.genbankfixer.gbtext_clean"><code class="xref py py-func docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a> function which can clean up broken Genbank files enough to
pass the BioPython Genbank parser</p>
<p>Almost all of this code was lifted from BioJSON (<a class="reference external" href="https://github.com/levskaya/BioJSON">https://github.com/levskaya/BioJSON</a>) by Anselm Levskaya.
The original code was not accompanied by any software licence. This parser is based on pyparsing.</p>
<p>There are some modifications to deal with fringe cases.</p>
<p>The parser first produces JSON as an intermediate format which is then formatted back into a
string in Genbank format.</p>
<p>The parser is not complete, so some fields do not survive the roundtrip (see below).
This should not be a difficult fix. The returned result has two properties,
.jseq which is the intermediate JSON produced by the parser and .gbtext
which is the formatted genbank string.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.parseGBLoc">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">parseGBLoc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#parseGBLoc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.parseGBLoc" title="Link to this definition"></a></dt>
<dd><p>retwingles parsed genbank location strings, assumes no joins of RC and FWD sequences</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.strip_multiline">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">strip_multiline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#strip_multiline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.strip_multiline" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toInt">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toInt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toInt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toInt" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.strip_indent">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">strip_indent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#strip_indent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.strip_indent" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.concat_dict">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">concat_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dlist</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#concat_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.concat_dict" title="Link to this definition"></a></dt>
<dd><p>more or less dict(list of string pairs) but merges
vals with the same keys so no duplicates occur</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toJSON">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toJSON</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gbkstring</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toJSON"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toJSON" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.wrapstring">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">wrapstring</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rowstart</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rowend</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padfirst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#wrapstring"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.wrapstring" title="Link to this definition"></a></dt>
<dd><p>wraps the provided string in lines of length rowend-rowstart
and padded on the left by rowstart.
-&gt; if padfirst is false the first line is not padded</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.locstr">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">locstr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#locstr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.locstr" title="Link to this definition"></a></dt>
<dd><p>genbank formatted location string, assumes no join’d combo of rev and fwd seqs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.originstr">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">originstr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#originstr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.originstr" title="Link to this definition"></a></dt>
<dd><p>formats dna sequence as broken, numbered lines ala genbank</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toGB">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toGB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jseq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toGB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toGB" title="Link to this definition"></a></dt>
<dd><p>parses json jseq data and prints out ApE compatible genbank</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.gbtext_clean">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">gbtext_clean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gbtext</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#gbtext_clean"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.gbtext_clean" title="Link to this definition"></a></dt>
<dd><p>This function takes a string containing <strong>one</strong> genbank sequence
in Genbank format and returns a named tuple containing two fields,
the gbtext containing a string with the corrected genbank sequence and
jseq which contains the JSON intermediate.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013</span>
<span class="gp">... </span><span class="s1">DEFINITION  .</span>
<span class="gp">... </span><span class="s1">ACCESSION</span>
<span class="gp">... </span><span class="s1">VERSION</span>
<span class="gp">... </span><span class="s1">SOURCE      .</span>
<span class="gp">... </span><span class="s1">  ORGANISM  .</span>
<span class="gp">... </span><span class="s1">COMMENT</span>
<span class="gp">... </span><span class="s1">COMMENT     ApEinfo:methylated:1</span>
<span class="gp">... </span><span class="s1">ORIGIN</span>
<span class="gp">... </span><span class="s1">        1 aaa</span>
<span class="gp">... </span><span class="s1">//&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.readers</span> <span class="kn">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>  
<span class="go">/home/bjorn/anaconda3/envs/bjorn36/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1388: BiopythonParserWarning: Malformed LOCUS line found - is this correct?</span>
<span class="go">:&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013\n&#39;</span>
<span class="go">  &quot;correct?\n:%r&quot; % line, BiopythonParserWarning)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;/home/bjorn/python_packages/pydna/pydna/readers.py&quot;</span>, line <span class="m">48</span>, in <span class="n">read</span>
<span class="w">    </span><span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="gr">IndexError</span>: <span class="n">pop from empty list</span>

<span class="gt">During handling of the above exception, another exception occurred:</span>

<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/home/bjorn/python_packages/pydna/pydna/readers.py&quot;</span>, line <span class="m">50</span>, in <span class="n">read</span>
<span class="w">    </span><span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No sequences found in data:</span><span class="se">\n</span><span class="s2">(</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">data</span><span class="p">[:</span><span class="mi">79</span><span class="p">]))</span>
<span class="gr">ValueError</span>: <span class="n">No sequences found in data:</span>
<span class="x">(LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013</span>
<span class="x">DEFINITI)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.genbankfixer</span> <span class="kn">import</span> <span class="n">gbtext_clean</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span><span class="p">,</span> <span class="n">j2</span> <span class="o">=</span> <span class="n">gbtext_clean</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
<span class="go">LOCUS       New_DNA                    3 bp ds-DNA     circular SYN 19-JUN-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION</span>
<span class="go">VERSION</span>
<span class="go">SOURCE      .</span>
<span class="go">ORGANISM  .</span>
<span class="go">COMMENT</span>
<span class="go">COMMENT     ApEinfo:methylated:1</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s3</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s3</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s3</span><span class="o">.</span><span class="n">format</span><span class="p">())</span>
<span class="go">LOCUS       New_DNA                    3 bp    DNA     circular SYN 19-JUN-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   New_DNA</span>
<span class="go">VERSION     New_DNA</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">COMMENT</span>
<span class="go">            ApEinfo:methylated:1</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.genbankrecord">
<span id="pydna-genbankrecord-module"></span><h2>pydna.genbankrecord module<a class="headerlink" href="#module-pydna.genbankrecord" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbankrecord.</span></span><span class="sig-name descname"><span class="pre">GenbankRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.from_string" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.from_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.pydna_code">
<span class="sig-name descname"><span class="pre">pydna_code</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.pydna_code"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.pydna_code" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.biopython_code">
<span class="sig-name descname"><span class="pre">biopython_code</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.biopython_code"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.biopython_code" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.myprimers">
<span id="pydna-myprimers-module"></span><h2>pydna.myprimers module<a class="headerlink" href="#module-pydna.myprimers" title="Link to this heading"></a></h2>
<p>Provides a practical way to access a list of primer sequences in a text file.</p>
<p>The path of a text file can be specified in the pydna.ini file or by the
´pydna_primers´ environment variable.</p>
<p>The file is expected to contain sequences in FASTA, Genbank or EMBL formats or
any format readable by the parse_primers function.</p>
<p>The primer list is expected to follow the convension below. The primer name is
expected to begin with the number.</p>
<p>can have the format below for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="mi">2</span><span class="n">_third_primer</span>
<span class="n">tgagtagtcgtagtcgtcgtat</span>

<span class="o">&gt;</span><span class="mi">1</span><span class="n">_second_primer</span>
<span class="n">tgatcgtcatgctgactatactat</span>

<span class="o">&gt;</span><span class="mi">0</span><span class="n">_first_primer</span>
<span class="n">ctaggatcgtagatctagctg</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The primerlist funtion returns a list of <a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.primer.Primer</span></code></a> objects
primerdict returns a dict where the key is the id of the object.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.myprimers.</span></span><span class="sig-name descname"><span class="pre">PrimerList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initlist:</span> <span class="pre">~typing.Iterable</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path:</span> <span class="pre">(&lt;class</span> <span class="pre">'str'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">&lt;class</span> <span class="pre">'pathlib.Path'&gt;)</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">&quot;p&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#PrimerList"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.PrimerList" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/collections.html#collections.UserList" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">UserList</span></code></a></p>
<p>Read a text file with primers.</p>
<p>The primers can be of any format readable by the parse_primers
function. Lines beginning with # are ignored. Path defaults to
the path given by the pydna_primers environment variable.</p>
<p>The primer list does not accept new primers. Use the
assign_numbers_to_new_primers method and paste the new
primers at the top of the list.</p>
<p>The primer list remembers the numbers of accessed primers.
The indices of accessed primers are stored in the .accessed
property.</p>
<dl class="py property">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList.accessed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">accessed</span></span><a class="headerlink" href="#pydna.myprimers.PrimerList.accessed" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList.assign_numbers">
<span class="sig-name descname"><span class="pre">assign_numbers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#PrimerList.assign_numbers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.PrimerList.assign_numbers" title="Link to this definition"></a></dt>
<dd><p>Find new primers in lst.</p>
<p>Returns a string containing new primers with their assigned
numbers. This string can be copied and pasted to the primer
text file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList.pydna_code_from_list">
<span class="sig-name descname"><span class="pre">pydna_code_from_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#PrimerList.pydna_code_from_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.PrimerList.pydna_code_from_list" title="Link to this definition"></a></dt>
<dd><p>Pydna code for a list of primer objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList.open_folder">
<span class="sig-name descname"><span class="pre">open_folder</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#PrimerList.open_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.PrimerList.open_folder" title="Link to this definition"></a></dt>
<dd><p>Open folder where primer file is located.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.myprimers.PrimerList.code">
<span class="sig-name descname"><span class="pre">code</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.myprimers.PrimerList.code" title="Link to this definition"></a></dt>
<dd><p>Pydna code for a list of primer objects.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.myprimers.check_primer_numbers">
<span class="sig-prename descclassname"><span class="pre">pydna.myprimers.</span></span><span class="sig-name descname"><span class="pre">check_primer_numbers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#check_primer_numbers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.check_primer_numbers" title="Link to this definition"></a></dt>
<dd><p>Find primers whose number do not match position in list.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.myprimers.undefined_sequence">
<span class="sig-prename descclassname"><span class="pre">pydna.myprimers.</span></span><span class="sig-name descname"><span class="pre">undefined_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#undefined_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.undefined_sequence" title="Link to this definition"></a></dt>
<dd><p>Primers in list with N or n instead of a sequence.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.myprimers.find_duplicate_primers">
<span class="sig-prename descclassname"><span class="pre">pydna.myprimers.</span></span><span class="sig-name descname"><span class="pre">find_duplicate_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#find_duplicate_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.myprimers.find_duplicate_primers" title="Link to this definition"></a></dt>
<dd><p>Find a list of lists with duplicated primer sequences.</p>
</dd></dl>

</section>
<section id="module-pydna.parsers">
<span id="pydna-parsers-module"></span><h2>pydna.parsers module<a class="headerlink" href="#module-pydna.parsers" title="Link to this heading"></a></h2>
<p>Provides two functions, parse and parse_primers</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.extract_from_text">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">extract_from_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#extract_from_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.extract_from_text" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.embl_gb_fasta">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">embl_gb_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#embl_gb_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.embl_gb_fasta" title="Link to this definition"></a></dt>
<dd><p>Parse embl, genbank or fasta format from text.</p>
<p>Returns list of Bio.SeqRecord.SeqRecord</p>
<p>annotations[“molecule_type”]
annotations[“topology”]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.parse">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#parse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.parse" title="Link to this definition"></a></dt>
<dd><p>Return <em>all</em> DNA sequences found in data.</p>
<p>If no sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em><em> or </em><em>iterable</em>) – <p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li><p>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences. Can be
a string or a Path object.</p></li>
<li><p>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</p></li>
<li><p>data can be a list or other iterable where the elements are 1 or 2</p></li>
</ol>
</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – If True double stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects are returned.
If False single stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> <a class="footnote-reference brackets" href="#id14" id="id13" role="doc-noteref"><span class="fn-bracket">[</span>7<span class="fn-bracket">]</span></a> objects are
returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains Dseqrecord or SeqRecord objects</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id14" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">7</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.parse_primers">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">parse_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#parse_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.parse_primers" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.primer">
<span id="pydna-primer-module"></span><h2>pydna.primer module<a class="headerlink" href="#module-pydna.primer" title="Link to this heading"></a></h2>
<p>This module provide the Primer class that is a subclass of the biopython SeqRecord.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.primer.Primer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.primer.</span></span><span class="sig-name descname"><span class="pre">Primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amplicon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">footprint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/primer.html#Primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.primer.Primer" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>Primer and its position on a template, footprint and tail.</p>
<dl class="py property">
<dt class="sig sig-object py" id="pydna.primer.Primer.footprint">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">footprint</span></span><a class="headerlink" href="#pydna.primer.Primer.footprint" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.primer.Primer.tail">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tail</span></span><a class="headerlink" href="#pydna.primer.Primer.tail" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.primer.Primer.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/primer.html#Primer.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.primer.Primer.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.readers">
<span id="pydna-readers-module"></span><h2>pydna.readers module<a class="headerlink" href="#module-pydna.readers" title="Link to this heading"></a></h2>
<p>Provides two functions, read and read_primer.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.readers.read">
<span class="sig-prename descclassname"><span class="pre">pydna.readers.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/readers.html#read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.readers.read" title="Link to this definition"></a></dt>
<dd><p>This function is similar the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code> function but expects one and only
one sequence or and exception is thrown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em>) – see below</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Double stranded or single stranded DNA, if True return
Dseqrecord objects, else Bio.SeqRecord objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains the first Dseqrecord or SeqRecord object parsed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The data parameter is similar to the data parameter for <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.readers.read_primer">
<span class="sig-prename descclassname"><span class="pre">pydna.readers.</span></span><span class="sig-name descname"><span class="pre">read_primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/readers.html#read_primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.readers.read_primer" title="Link to this definition"></a></dt>
<dd><p>Use this function to read a primer sequence from a string or a local file.
The usage is similar to the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse_primer()</span></code> function.</p>
</dd></dl>

</section>
<section id="module-pydna.seqrecord">
<span id="pydna-seqrecord-module"></span><h2>pydna.seqrecord module<a class="headerlink" href="#module-pydna.seqrecord" title="Link to this heading"></a></h2>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></p>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.from_Bio_SeqRecord">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_Bio_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SeqRecord</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.from_Bio_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord" title="Link to this definition"></a></dt>
<dd><p>Creates a pydnaSeqRecord from a Biopython SeqRecord.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.locus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">locus</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.locus" title="Link to this definition"></a></dt>
<dd><p>Alias for name property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.accession">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">accession</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.accession" title="Link to this definition"></a></dt>
<dd><p>Alias for id property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.definition">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">definition</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.definition" title="Link to this definition"></a></dt>
<dd><p>Alias for description property.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rc" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.isorf" title="Link to this definition"></a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id16" id="id15" role="doc-noteref"><span class="fn-bracket">[</span>8<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id16" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">8</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.translate" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape">
<span class="sig-name descname"><span class="pre">add_colors_to_features_for_ape</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.add_colors_to_features_for_ape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape" title="Link to this definition"></a></dt>
<dd><p>Assign colors to features.</p>
<p>compatible with
the <a class="reference external" href="http://jorgensen.biology.utah.edu/wayned/ape/">ApE editor</a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_feature" title="Link to this definition"></a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2),</span>
<span class="go">                           ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;,</span>
<span class="go">            qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.list_features">
<span class="sig-name descname"><span class="pre">list_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.list_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.list_features" title="Link to this definition"></a></dt>
<dd><p>Print ASCII table with all features.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">list_features</span><span class="p">())</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">| Ft# | Label or Note | Dir | Sta | End | Len | type | orf? |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">|   0 | L:ft2         | --&gt; | 2   | 4   |   2 | misc |  no  |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.extract_feature" title="Link to this definition"></a></dt>
<dd><p>Extract feature and return a new SeqRecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>)</p></li>
<li><p><strong>extract</strong> (<em>Indicates the feature to</em>)</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">SeqRecord(seq=Seq(&#39;gt&#39;), id=&#39;ft2&#39;, name=&#39;part_name&#39;,</span>
<span class="go">          description=&#39;description&#39;, dbxrefs=[])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.sorted_features">
<span class="sig-name descname"><span class="pre">sorted_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.sorted_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.sorted_features" title="Link to this definition"></a></dt>
<dd><p>Return a list of the features sorted by start position.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">sorted_features</span><span class="p">())</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.seguid" title="Link to this definition"></a></dt>
<dd><p>Return the url safe SEGUID <a class="footnote-reference brackets" href="#id18" id="id17" role="doc-noteref"><span class="fn-bracket">[</span>9<span class="fn-bracket">]</span></a> for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base 64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be a part of and URL or a filename.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;gattaca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span> <span class="c1"># original seguid is +bKGnebMkia5kNg/gF7IORXMnIU</span>
<span class="go">&#39;lsseguid=tp2jzeCM2e3W4yxtrrx09CMKa_8&#39;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id18" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">9</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.comment">
<span class="sig-name descname"><span class="pre">comment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">newcomment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.comment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.comment" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.datefunction">
<span class="sig-name descname"><span class="pre">datefunction</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.datefunction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.datefunction" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stamp">
<span class="sig-name descname"><span class="pre">stamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">now</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">datefunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pydna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'</span> <span class="pre">'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.stamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stamp" title="Link to this definition"></a></dt>
<dd><p>Add seguid checksum to COMMENTS sections</p>
<p>The checksum is stored in object.annotations[“comment”].
This shows in the COMMENTS section of a formatted genbank file.</p>
<p>For blunt linear sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">SEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>For circular sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">cSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>Fore linear sequences which are not blunt:</p>
<p><code class="docutils literal notranslate"><span class="pre">lSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="go">&#39;lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttjU&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">][:</span><span class="mi">41</span><span class="p">]</span>
<span class="go">&#39;pydna lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttj&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.lcs">
<span class="sig-name descname"><span class="pre">lcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.lcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.lcs" title="Link to this definition"></a></dt>
<dd><p>Return the longest common substring between the sequence.</p>
<p>and another sequence (other). The other sequence can be a string,
Seq, SeqRecord, Dseq or DseqRecord.
The method returns a SeqFeature with type “read” as this method
is mostly used to map sequence reads to the sequence. This can be
changed by passing a type as keyword with some other string value.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(0),</span>
<span class="go">                          ExactPosition(6), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(1),</span>
<span class="go">                          ExactPosition(5), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;CCCCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(None)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.gc" title="Link to this definition"></a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.cai" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rarecodons" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.startcodon" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stopcodon" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.express" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.copy" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protocol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.dump" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">ProteinSeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rc" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.isorf" title="Link to this definition"></a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id20" id="id19" role="doc-noteref"><span class="fn-bracket">[</span>10<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id20" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id19">10</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.gc" title="Link to this definition"></a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.cai" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.startcodon" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#ProteinSeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.express" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.tm">
<span id="pydna-tm-module"></span><h2>pydna.tm module<a class="headerlink" href="#module-pydna.tm" title="Link to this heading"></a></h2>
<p>This module provide functions for melting temperature calculations.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_default">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.DNA_NN4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">de_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span> <span class="pre">/</span> <span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span> <span class="pre">/</span> <span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selfcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tris</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">75.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dNTPs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltcorr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.Tm_NN</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tm_default"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_default" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_dbd">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_dbd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.DNA_NN3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">de_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selfcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tris</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dNTPs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltcorr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.Tm_NN</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tm_dbd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_dbd" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_product">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_product</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.050</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tm_product"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_product" title="Link to this definition"></a></dt>
<dd><p>Tm calculation for the amplicon.</p>
<p>according to:</p>
<p>Rychlik, Spencer, and Rhoads, 1990, Optimization of the anneal
ing temperature for DNA amplification in vitro
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/2243783">http://www.ncbi.nlm.nih.gov/pubmed/2243783</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.ta_default">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">ta_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_product</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_product</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#ta_default"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.ta_default" title="Link to this definition"></a></dt>
<dd><p>Ta calculation.</p>
<p>according to:</p>
<p>Rychlik, Spencer, and Rhoads, 1990, Optimization of the anneal
ing temperature for DNA amplification in vitro
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/2243783">http://www.ncbi.nlm.nih.gov/pubmed/2243783</a></p>
<p>The formula described uses the length and GC content of the product and
salt concentration (monovalent cations)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.ta_dbd">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">ta_dbd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_product</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#ta_dbd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.ta_dbd" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_default</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.program" title="Link to this definition"></a></dt>
<dd><p>Returns a string containing a text representation of a suggested
PCR program using Taq or similar polymerase.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|95°C|95°C               |    |tmf:59.5
|____|_____          72°C|72°C|tmr:59.7
|3min|30s  \ 59.1°C _____|____|60s/kb
|    |      \______/ 0:32|5min|GC 51%
|    |       30s         |    |1051bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.taq_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">taq_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_default</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.tm.taq_program" title="Link to this definition"></a></dt>
<dd><p>Returns a string containing a text representation of a suggested
PCR program using Taq or similar polymerase.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|95°C|95°C               |    |tmf:59.5
|____|_____          72°C|72°C|tmr:59.7
|3min|30s  \ 59.1°C _____|____|60s/kb
|    |      \______/ 0:32|5min|GC 51%
|    |       30s         |    |1051bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.dbd_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">dbd_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_dbd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#dbd_program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.dbd_program" title="Link to this definition"></a></dt>
<dd><p>Text representation of a suggested PCR program.</p>
<p>Using a polymerase with a DNA binding domain such as Pfu-Sso7d.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|98°C|98°C               |    |tmf:53.8
|____|_____          72°C|72°C|tmr:54.8
|30s |10s  \ 57.0°C _____|____|15s/kb
|    |      \______/ 0:15|5min|GC 51%
|    |       10s         |    |1051bp

|98°C|98°C      |    |tmf:82.5
|____|____      |    |tmr:84.4
|30s |10s \ 72°C|72°C|15s/kb
|    |     \____|____|GC 52%
|    |      3:45|5min|15058bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.pfu_sso7d_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">pfu_sso7d_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_dbd</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.tm.pfu_sso7d_program" title="Link to this definition"></a></dt>
<dd><p>Text representation of a suggested PCR program.</p>
<p>Using a polymerase with a DNA binding domain such as Pfu-Sso7d.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|98°C|98°C               |    |tmf:53.8
|____|_____          72°C|72°C|tmr:54.8
|30s |10s  \ 57.0°C _____|____|15s/kb
|    |      \______/ 0:15|5min|GC 51%
|    |       10s         |    |1051bp

|98°C|98°C      |    |tmf:82.5
|____|____      |    |tmr:84.4
|30s |10s \ 72°C|72°C|15s/kb
|    |     \____|____|GC 52%
|    |      3:45|5min|15058bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.Q5">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">Q5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#Q5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.Q5" title="Link to this definition"></a></dt>
<dd><p>For Q5 Ta they take the lower of the two Tms and add 1C
(up to 72C). For Phusion they take the lower of the two
and add 3C (up to 72C).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tmbresluc">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tmbresluc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primerc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tmbresluc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tmbresluc" title="Link to this definition"></a></dt>
<dd><p>Returns the tm for a primer using a formula adapted to polymerases
with a DNA binding domain, such as the Phusion polymerase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>primer</strong> (<em>string</em>) – primer sequence 5’-3’</p></li>
<li><p><strong>primerc</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – primer concentration in nM), set to 500.0 nm by default.</p></li>
<li><p><strong>saltc</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a><em>, </em><em>optional</em>) – Monovalent cation concentration in mM, set to 50.0 mM by default.</p></li>
<li><p><strong>thermodynamics</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – prints details of the thermodynamic data to stdout. For
debugging only.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tm</strong> – the tm of the primer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_neb">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_neb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prodcode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'q5-0'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tm_neb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_neb" title="Link to this definition"></a></dt>
<dd><p>Calculates a single primers melting temp from NEB.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>primer1</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>conc</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>)</p></li>
<li><p><strong>prodcode</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – find product codes on nebswebsite: <a class="reference external" href="https://tmapi.neb.com/docs/productcodes">https://tmapi.neb.com/docs/productcodes</a></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tm</strong> – primer melting temperature</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-pydna.utils">
<span id="pydna-utils-module"></span><h2>pydna.utils module<a class="headerlink" href="#module-pydna.utils" title="Link to this heading"></a></h2>
<p>Miscellaneous functions.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.three_frame_orfs">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">three_frame_orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startcodons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('ATG',)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stopcodons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('TAG',</span> <span class="pre">'TAA',</span> <span class="pre">'TGA')</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#three_frame_orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.three_frame_orfs" title="Link to this definition"></a></dt>
<dd><p>Overlapping orfs in three frames.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.shift_location">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">shift_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">original_location</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#shift_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.shift_location" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.shift_feature">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">shift_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#shift_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.shift_feature" title="Link to this definition"></a></dt>
<dd><p>Return a new feature with shifted location.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.smallest_rotation">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">smallest_rotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#smallest_rotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.smallest_rotation" title="Link to this definition"></a></dt>
<dd><p>Smallest rotation of a string.</p>
<p>Algorithm described in Pierre Duval, Jean. 1983. Factorizing Words
over an Ordered Alphabet. Journal of Algorithms &amp; Computational Technology
4 (4) (December 1): 363–381. and Algorithms on strings and sequences based
on Lyndon words, David Eppstein 2011.
<a class="reference external" href="https://gist.github.com/dvberkel/1950267">https://gist.github.com/dvberkel/1950267</a></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.utils</span> <span class="kn">import</span> <span class="n">smallest_rotation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">smallest_rotation</span><span class="p">(</span><span class="s2">&quot;taaa&quot;</span><span class="p">)</span>
<span class="go">&#39;aaat&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.cai">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">_weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.cai" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.rarecodons">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.rarecodons" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.express">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.express" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
<p><strong>NOT IMPLEMENTED YET</strong></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.open_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">open_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pth</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#open_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.open_folder" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.rc">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">StrOrBytes</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">StrOrBytes</span></span></span><a class="reference internal" href="_modules/pydna/utils.html#rc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.rc" title="Link to this definition"></a></dt>
<dd><p>Reverse complement.</p>
<p>accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.complement">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.complement" title="Link to this definition"></a></dt>
<dd><p>Complement.</p>
<p>accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.memorize">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">memorize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#memorize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.memorize" title="Link to this definition"></a></dt>
<dd><p>Cache functions and classes.</p>
<p>see pydna.download</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.identifier_from_string">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">identifier_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="_modules/pydna/utils.html#identifier_from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.identifier_from_string" title="Link to this definition"></a></dt>
<dd><p>Return a valid python identifier.</p>
<p>based on the argument s or an empty string</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.flatten">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">flatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a></span></span><a class="reference internal" href="_modules/pydna/utils.html#flatten"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.flatten" title="Link to this definition"></a></dt>
<dd><p>Flattens an iterable of iterables.</p>
<p>Down to str, bytes, bytearray or any of the pydna or Biopython seq objects</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.seq31">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">seq31</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#seq31"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.seq31" title="Link to this definition"></a></dt>
<dd><p>Turn a three letter code protein sequence into one with one letter code.</p>
<p>The single input argument ‘seq’ should be a protein sequence using single
letter codes, as a python string.</p>
<p>This function returns the amino acid sequence as a string using the one
letter amino acid codes. Output follows the IUPAC standard (including
ambiguous characters B for “Asx”, J for “Xle” and X for “Xaa”, and also U
for “Sel” and O for “Pyl”) plus “Ter” for a terminator given as an
asterisk.</p>
<p>Any unknown
character (including possible gap characters), is changed into ‘Xaa’.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">seq3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGAR*&quot;</span><span class="p">)</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.utils</span> <span class="kn">import</span> <span class="n">seq31</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq31</span><span class="p">(</span><span class="s1">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span><span class="p">)</span>
<span class="go">&#39;M  A  I  V  M  G  R  W  K  G  A  R  *&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomRNA">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#randomRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomRNA" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomDNA">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#randomDNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomDNA" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomORF">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomORF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#randomORF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomORF" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomprot">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomprot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#randomprot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomprot" title="Link to this definition"></a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.eq">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">eq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#eq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.eq" title="Link to this definition"></a></dt>
<dd><p>Compare two or more DNA sequences for equality.</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same double stranded DNA molecule.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable</em>) – iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>eq</strong> – Returns True or False</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ol class="arabic simple">
<li><p>They have the same sequence (case insensitive)</p></li>
<li><p>One sequence is the reverse complement of the other</p></li>
</ol>
<p>Two circular sequences are considered equal if they are circular
permutations meaning that they have the same length and:</p>
<ol class="arabic simple">
<li><p>One sequence can be found in the concatenation of the other sequence with itself.</p></li>
<li><p>The reverse complement of one sequence can be found in the concatenation of the other sequence with itself.</p></li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.utils</span> <span class="kn">import</span> <span class="n">eq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.cuts_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">cuts_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#cuts_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.cuts_overlap" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.location_boundaries">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">location_boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SimpleLocation</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">CompoundLocation</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#location_boundaries"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.location_boundaries" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.locations_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">locations_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loc1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SimpleLocation</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">CompoundLocation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SimpleLocation</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">CompoundLocation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#locations_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.locations_overlap" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Link to this heading"></a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Björn F. Johansson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>